object bucket provisioning (#1383)
kubernetes/enhancements: keps/sig-storage/20191125-bucket-provisioning.md

---
title: Object Bucket Provisioning
authors:
  - "@jeffvance"
  - "@copejon"
owning-sig: "sig-storage"
reviewers:
  - "@saad-ali"
  - "@alarge"
  - "@erinboyd"
  - "@guymguym"
  - "@travisn"
approvers:
  - TBD
editor: TBD
creation-date: 2019-11-25
last-updated: 2020-02-27
status: provisional
---

# Object Bucket Provisioning

## Table of Contents

<!-- toc -->
- [Summary](#summary)
  - [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
  - [Vocabulary](#vocabulary)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
      - [Admin](#admin)
      - [User](#user)
  - [System Configuration](#system-configuration)
  - [Workflows](#workflows)
    - [Create Bucket](#create-bucket)
    - [Delete Bucket](#delete-bucket)
    - [Grant Bucket Access](#grant-bucket-access)
    - [Revoke Bucket Access](#revoke-bucket-access)
  - [Custom Resource Definitions](#custom-resource-definitions)
      - [ObjectBucketClaim (OBC)](#objectbucketclaim-obc)
      - [ObjectBucket (OB)](#objectbucket-ob)
      - [BucketClass](#bucketclass)
<!-- /toc -->

# Summary

> This proposal introduces the Container Object Storage Interface (COSI), whose purpose is to provide a familiar and standardized method of provisioning object storage buckets in an manner agnostic to the storage vendor. The COSI environment is comprised of Kubernetes CRDs, operators to manage these CRDs, and a gRPC interface by which these operators communicate with vendor plugins.  This design is of course heavily inspired by the Kubernetes’ implementation of the Container Storage Interface (CSI).  However, bucket management lacks some of the notable requirements of block and file provisioning and allows for an architecture with lower overall complexity.  This proposal does not include a standardized protocol or abstraction of storage vendor APIs.  

Reviewer: It has nothing to do with Containers. So I think Container Object Storage Interface may not be the right term.

## Motivation 

File and block are first class citizens within the Kubernetes ecosystem.  Object, though different on a fundamental level and lacking an open, committee controlled interface like POSIX, is a popular means of storing data, especially against very large data sources.   As such, we feel it is in the interest of the community to elevate buckets to a community supported feature.  In doing so, we can provide Kubernetes cluster users and administrators a normalized and familiar means of managing object storage.

## Goals
+ Define a control plane API in order to standardize and formalize a Kubernetes bucket provisioning
+ Minimize privileges required to run a storage driver.
+ Minimize technical ramp-up for storage vendors.
+ Be un-opinionated about the underlying object-store.
+ Use Kubernetes Secrets to inject bucket information into application workflows.
+ Present similar workflows for both _greenfield_  and _brownfield_ bucket provisioning.
+ Present a design that is familiar to CSI storage plugin authors and Kubernetes storage admins.

## Non-Goals

> + Define a native _data-plane_ object store API which would greatly improve object store app portability.

Reviewer: Not sure how to handle this in general. The main difference then between block/filesystem provisioning and this is in block/filesystem the user is abstracted away from knowing what protocol is used to talk to the backing store, how to authenticate with the backing store if required, and even needing any kind of driver at all to talk to it. K8s itself provides the container with just a mount.

If an abstract dataplane is not provided, then all those details must be part of the api itself and left up as an implementation detail to the app, making it potentially not portable.

It still may be worth doing though.

Reviewer 2: This has been debated for a long time, and in fact has a pretty rich startup space, too .  Providing a universal data plane API means either a) choosing a lowest-common API (a subset of S3 seems most widely plausible) or injecting an actor into the datapath (which is scary from an auth and perf and reliability POV).

Alternatively, let this be a client-side thing.  Make client libs that understand backends natively.  Leave this API for "creations, credentials, and coordinates".

Reviewer 3: @kfox1111 +1 As Tim pointed out, the idea has kicked around for a while.  One of the motivations for this KEP is to lay some groundwork for a k8s native object interface.

Author: I would like to work on a data plane object KEP, but after this KEP is (hopefully) merged and a code pr is underway.

+ _Static_ - also called "static brownfield" - buckets are created outside the COSI system but granted access via COSI without the need for the driver/sidecar. The OB and driver secret are created manually in the expected locations. The app secret and binding are performed by COSI.

# Proposal

## User Stories

#### Admin

- As a cluster administrator, I can set quotas and resource limits on generated buckets' storage capacity via the Kubernete's API, so that  I can control monthly infrastructure costs.
- As a cluster administrator, I can use Kubernetes RBAC policies on bucket APIs, so that I may control integration and access to pre-existing buckets from within the cluster, reducing the need to administer an external storage interface.
- As a cluster administrator, I can manage multiple object store providers via the Kubernetes API, so that I do not have to become an expert in several different storage interfaces.

#### User

- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.

> - As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.

Reviewer 3: It's expected that Pods will mount secrets containing the credential and endpoint information.  these secrets are generated by the COSI controller.

> - As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.

Author: Current design does not rely on secrets or config maps. Instead, as CSI volume is defined in the app pod. The kubelet pauses the launch of the pod until the target adapter, via gRPC, tells the kubelet storage has been handled.

  
## System Configuration

+ The central controller runs in a protected namespace with RBAC privileges for managing OBCs, OBs, and Secrets cluster wide.
    + Responsible for the binding relationship of OBs and OBCs.
+ The "cosi-system" namespace must be created.

>     + This namespace must be granted write access to cluster-wide scoped OBs.

Reviewer 3: +1

>     + This namespace must be granted write access to cluster-wide scoped OBs.

Author: a later iteration renames OB -> `BucketContent` and it is cluster scoped.

>     + This namespace must be granted write access to cluster-wide scoped OBs.

Reviewer 4: well, that answers from up above.  So both BucketContent and BucketClass are non-namespaced?

> + The Driver and Sidecar containers run together in a single Pod in the "cosi-system" namespace.

Reviewer 3: >  I'd prefer one driver = one namespace. 

I see the pro's and con's of this.  RBAC gets a little messier (and more prone to opening security holes) with 1:1 namespace:driver.  This is reduced by making the OB and BucketClass cluster scoped - there's no longer a concern that a user could create their own OB and/or BC.

Encapsulation of a driver in a ns with only it's managed resources (OBs/Secrets) provides for a nice security story around compromised drivers.

Ultimately though, unless we hard-code the `cosi-system` namespace, the topology would be something prescribed in documentation.  Admins would essentially be free to use a single, grouped, or discrete namespaces.

> + The Driver and Sidecar containers run together in a single Pod in the "cosi-system" namespace.

Reviewer 4: I'm less worried about the namespace a driver would run in than I am about where the OB class and bucket content live.  If those two have a single well-known namespace, I'm ambivalent about driver namespace.

    + The gRPC connection is made through the Pod’s localhost.
+ BucketClasses must exist (cluster-wide scope).
+ No node affinity or other requirements exist.
+ Operations must be idempotent in order to handle failure recovery.
  

> ## Workflows

Reviewer 3: +1 will addr

### Create Bucket
1. The user creates an OBC in the app’s namespace.
1. The COSI central controller detects the OBC and creates an OB (cluster-wide), containing driver-relevant data collected from the OBC and BC.
1. The Sidecar detects the OB and issues a CreateBucket() rpc, passing to the driver config data.
1. The Driver creates the bucket and returns pertinent endpoint, credential, and metadata.
1. The Sidecar creates a secret in its namespace ("cosi-system") containing essential connection information.

> 1. The central controller copies the secret to the OBC’s namespace and the app is ready to run.

Reviewer 3: > the source of the credentials is not a Secrets objects in user namespace, but Secrets (or any other object...) in bucket driver namespace.

Could this be something like the `CSIPersistentVolumeSource` wherein the driver-namespace Secret is referenced? Or indirectly referenced via the `BucketContent`? 

And could we bridge the gap between mounted files and env vars with application quiescence/un-quiescence execution hooks?  They were mentioned in the first review meeting as potentially enabling something like that.

Avoiding having to copy the secret would be great if we can manage it.

### Revoke Bucket Access (needs to handle stratic and dynamic brownfield!)

Reminder: BucketClass is ignored in browfield operations.

1. The user deletes the OBC, which blocks until the finalizer is removed.
1. The central controller detects the OBC change and updates OB’s Phase.
1. The Sidecar detects the OB update and calls the `Revoke()` gRPC method.
1. The Driver revokes access to the bucket and responds with no errors.
1. The Sidecar deletes the OB’s secret.
1. The central controller deletes the secret’s copy in the app’s namespace.
1. The central controller removes the finalizers from the OBC, OB, and Secret.
1. The OBC and its secret are garbage collected and the OB remains.

##  Custom Resource Definitions

#### ObjectBucketClaim (OBC)

A user facing API object representing a request for a bucket, or access to an existing bucket. OBCs are created by users in their namespaces. Once the request is fulfilled, the OBC is “bound” to an Object Bucket (OB). The binding is used to mark the request as fulfilled and prevent further binds to the OB.

```yaml
apiVersion: cosi.io/v1alpha1

> kind: ObjectBucketClaim

Reviewer 3: +1

> kind: ObjectBucketClaim

Author: We've added protocols to the `BucketClass` and to the `Bucket` (formerly OBC).

metadata:
  name: [1]
  namespace:
  labels:
    “cosi.io/driver”: [2]
  finalizers:
  - cosi.io/finalizer [3]
spec:
  bucketName: [4]

>   generateBucketName: [5]

Reviewer 3: Buckets may be named and that name can be used to construct a public URL for that bucket.  We felt that that there may be cases where a user wants to name the buckets explicitly (with collision errors being returned by the driver and recorded on the Bucket object).  For users unconcerned with naming, a prefix may be given for a randomized name to guard against collisions.

>   generateBucketName: [5]

Reviewer 2: Alternative: if you want an explicit, static name, don't use the "greenfield" path.  Create the bucket yourself and use the brownfield path.

>   generateBucketName: [5]

Reviewer 3: @thockin This does mean that a user would have to be allowed to create ObjectBuckets in the cluster-scope to represent their object storage and Secerts in the driver's namespace.  That said, I see that being within the realm of the design - basically indistinguishable from other static buckets (provided they have the privs or get an admin to do it).

>   generateBucketName: [5]

Reviewer 2: I'm not sure how this *should* work, but the existing precedents ar einteresting.

Generally we encourage people to think of dynamic provisioning as ephemeral (IPs, PVs, etc) and non-ephemeral cases are handled out-of-band (e.g. manually allocate a block device, create a PV that links to it and pre-binds to a PVC, then use it from that PVC).

We don't have an equivalent for IPs (maybe we should), but I know many people let the IP be allocated dynamically by the cloud provider and then manually "promote" it to static by naming it.

In short, allowing users to create statically named things is kind of a DOS vector and handled OOB from the provisioning API.

>   generateBucketName: [5]

Author: > Create the bucket yourself and use the brownfield path.

I see the benefit of being able to create a random bucket name as increasing the chances of the new bucket's name being unique in a world of huge bucket namespaces. Instead, we could pass a flag to the driver saying "you generate a name".

>   generateBucketName: [5]

Reviewer 4: FWIW, I made a similar comment to Tim's a while ago - explicitly providing a full bucket name for a greenfield case just feels like it violates the model.  Prefix, sure.  But whole bucket name?

  bucketClassName: [6]
  objectBucketName: [7]
  secretName: [8]
  driverSecretName: [9]
status:
  phase:
  conditions: []ObjectBucketClaimCondition
```
1. `name`: the metadata name of the OBC; however this name can be generated and thus cannot be relied on to predictably name other related resources, eg. secrets. 
1. `labels`: central controller adds the label to its managed resources for easy GET ops.  Value is the driver name returned by GetDriverInfo() rpc*.
1. `finalizers`: central controller adds the finalizer to defer OBC deletion until backend deletion ops succeed.
1. `bucketName`: Desired name of the bucket to be created**.
1. `generateBucketName`: Prefix to a randomly generated name. Mutually exclusive with `bucketName`**.
1. `bucketClassName`: Name of the target BucketClass**.

> 1. `objectBucketName`: Name of a bound OB.

Reviewer 3: We've renamed `objectBucket` to `bucketContent`, which should help with confusion.

   - Injected by the central controller during greenfield ops.
   - Defined by the OBC author for static and dynamic brownfield ops.
1. `secretName`: Desired name of the app's secret for greenfield. Fails if secret exists. Defined here so that app deployment (where the secret name is required) is independent of OBC creation.
1. `driverSecretName`: Name of the driver's secret with the namespace assumed to be "cosi-system".
   - Injected by the central controller during greenfield and dynamic brownfield ops.

>    - Defined by the OBC author for static brownfield ops.

Author: `driverSecretName` has been removed in a later version. But the idea of guessing a provisioner's secret name in the `cosi-system` ns or in a provisioner's ns if we chose that, should be covered by RBAC (I think).

\* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

** Ignored for brownfield usage.

#### ObjectBucket (OB)
A cluster-wide scoped resource representing the bucket. The OB is expected to store stateful data relevant to bucket deprovisioning. The OB is bound to the OBC in a 1-1 mapping.

```yaml
apiVersion: cosi.io/v1alpha1
kind: ObjectBucket
Metadata:
  name: [1]
  namespace: [2]
  labels:
    “cosi.io/driver”: [3]
  finalizers:
  - "cosi.io/finalizer" [4]

> spec:

Author: Add: the `Bucket` contains all of the relevant info from the BucketClass at the time the Bucket is created.

> spec:

Author: @jsafrane I _think_  the idea of a handle was discussed, and due to the 1:many mapping of Bucket:BucketRequest, rather than a list of BucketRequests (which needs to be maintained), a handle is omitted. This means `kubectl describe bucket` won't show the BucketRequest that generated this Bucket.

  bucketClassName: [5]
  releasePolicy: {"Delete", "Retain"} [6]

>   bucketConfig: map[string]string [7]

Reviewer 3: +1

  objectBucketClaimRef: [8]
    name:
    namespace:
  secretName: [9]
status:
  bucketMetaData: <map[string]string> [10]
  phase: {"Bound", "Released", "Failed", “Errored”} [11]
  conditions: []ObjectBucketCondition
```

> 1. `name`: Generated in the pattern of “obc-”\<OBC-NAMESPACE>"-"\<OBC-NAME>

Reviewer 3: Agreed, though I remember the max being ~64 characters.  We used this naming scheme with our library incantation of this without issue.

> 1. `name`: Generated in the pattern of “obc-”\<OBC-NAMESPACE>"-"\<OBC-NAME>

Author: This is a good point.
[types](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L110) `name`. `namespace`, and `generateName` are all unbound _string_ types. It [appears](DNS1123LabelMaxLength) that the max name length is 63 chars, so I assume we have a problem if we add a 63 byte namespace to a 63 byte OBC name with a 4 byte prefix.

> 1. `name`: Generated in the pattern of “obc-”\<OBC-NAMESPACE>"-"\<OBC-NAME>

Reviewer 2: Deployment -> replicaset -> pod has the same problem.  It's not cleanly handled in a lot of places.  I encourage you to try to do better, but maybe don't die on this hill.

> 1. `name`: Generated in the pattern of “obc-”\<OBC-NAMESPACE>"-"\<OBC-NAME>

Author: In our API validation we could fail the obc creation if it's `name + ns > (63-PREFIX-LENGTH)`.

## Motivation 

File and block are first class citizens within the Kubernetes ecosystem.  Object, though different on a fundamental level and lacking an open, committee controlled interface like POSIX, is a popular means of storing data, especially against very large data sources.   As such, we feel it is in the interest of the community to elevate buckets to a community supported feature.  In doing so, we can provide Kubernetes cluster users and administrators a normalized and familiar means of managing object storage.

## Goals
+ Define a control plane API in order to standardize and formalize a Kubernetes bucket provisioning
+ Minimize privileges required to run a storage driver.

> + Minimize technical ramp-up for storage vendors.

Reviewer 2: Unlike CSI which has potentially hundreds of vendor-drivers, how many object-store drivers will really exist?  I just want to make sure we don't over-complicate the design "because CSI" when the goals may actually be pretty different.

Author: Way more than I thought.
In no particular order:
**vendors**: amazon, google, red hat (separate from ibm), ibm, emc, netapp, minio, microsoft, dell (separate from emc), swiftstack, cloudian, caringo, data direct networks, HGST, hitachi, huawei, pure, scality, western digital, qumulo, oracle, rackspace, digitalocean, alibaba, vultr, zadara, exablox, tarmin, gScale, oneblox. openIO, quantastor, igneous, basho, etergo, hyve, newisys, supermicro, seagate, hpe, 

**tech**: ceph rgw, aws s3, azure blobs, leofs, swift, minio, edgefs

There may be some dups. I suspect a lot of these are at least loosely based on s3, so an `s3-compatible` protocol would apply to many.

Reviewer 2: How many of those vendors will use the same driver (when they have same tech) vs have totally distinct drivers?

Author: I hadn't considered shared drivers but that's an interesting thought.
Although the current proposal states all drivers live in the _cosi-system_ ns, that is up for debate as there are opinions that drivers (and their sidecars) should be namespace isolated (likely in the same ns as the object store when it resides in the k8s cluster).
If this KEP is accepted sig-storage would likely support the COSI sidecar controller and my thought was that each vendor would write their own driver. Of course many of the `s3-compatible` drivers would be similar and could technically be shared. I supposed sig-storage could even support an `s3-compatible` driver and let vendors who want to differentiate themselves write their own?

Reviewer 4: While I expect that bucket creation and deletion might fold into a few uses cases, I'm not clear that would be the case for IAM-style parameters.  They way that identify is communicated to the bucket storage system is likely mostly standardized, but not how you actually setup  those identifies and define their access control.

  

## System Configuration

+ The central controller runs in a protected namespace with RBAC privileges for managing OBCs, OBs, and Secrets cluster wide.
    + Responsible for the binding relationship of OBs and OBCs.
+ The "cosi-system" namespace must be created.
    + This namespace must be granted write access to cluster-wide scoped OBs.
+ The Driver and Sidecar containers run together in a single Pod in the "cosi-system" namespace.
    + The gRPC connection is made through the Pod’s localhost.
+ BucketClasses must exist (cluster-wide scope).
+ No node affinity or other requirements exist.
+ Operations must be idempotent in order to handle failure recovery.
  
## Workflows

### Create Bucket
1. The user creates an OBC in the app’s namespace.
1. The COSI central controller detects the OBC and creates an OB (cluster-wide), containing driver-relevant data collected from the OBC and BC.
1. The Sidecar detects the OB and issues a CreateBucket() rpc, passing to the driver config data.
1. The Driver creates the bucket and returns pertinent endpoint, credential, and metadata.
1. The Sidecar creates a secret in its namespace ("cosi-system") containing essential connection information.

> 1. The central controller copies the secret to the OBC’s namespace and the app is ready to run.

Reviewer 2: Please make sure to consider credentialing mechanisms that don't need explicit secrets.  E.g. when the existing kubernetes ServiceAccount is sufficient to access a bucket.

This will require information like "which Kube SA" and will not need information like "which secret"

Reviewer 3: @thockin Thanks pointing this out - I hadn't realized that external services might use SAs as a means of authentication but it would simply workflow management.

Reviewer 2: Google Cloud's Workload Identity operates like this.  I don't want to bias the design to my employer's needs, but I expect it to be a more common pattern over time.

1. `labels`: central controller adds the label to its managed resources for easy GET ops.  Value is the driver name returned by GetDriverInfo() rpc*.
1. `finalizers`: central controller adds the finalizer to defer OBC deletion until backend deletion ops succeed.
1. `bucketName`: Desired name of the bucket to be created**.
1. `generateBucketName`: Prefix to a randomly generated name. Mutually exclusive with `bucketName`**.
1. `bucketClassName`: Name of the target BucketClass**.
1. `objectBucketName`: Name of a bound OB.
   - Injected by the central controller during greenfield ops.
   - Defined by the OBC author for static and dynamic brownfield ops.
1. `secretName`: Desired name of the app's secret for greenfield. Fails if secret exists. Defined here so that app deployment (where the secret name is required) is independent of OBC creation.
1. `driverSecretName`: Name of the driver's secret with the namespace assumed to be "cosi-system".
   - Injected by the central controller during greenfield and dynamic brownfield ops.
   - Defined by the OBC author for static brownfield ops.

> \* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

Reviewer 2: How do you deal with conflicts?

Author: Reasons for conflicts: 1) two drivers return the same name, 2) we replace offending driver name chars with "-" and create our own conflict.
The first case could be intentional. The second case would appear to be a bug. Is there a better "solution" than documenting legal driver names?

**Note**: if we go back to proposing that drivers are isolated by ns, then the value should include the ns but conflicts could still occur.

Reviewer 3: This could be handled by registering the drivers as they're installed and on COSI controller start.  

Registration could be accomplished in a couple ways e.g.:

Using a ProvisionerRegistration object to abstract the provisioner.  On ADD events, COSI verifies that the driver name isn't already taken.  

Or using a reverse facing gRPC interface implemented by the COSI controller.  Sidecars carry the client for this interface and on start, get the driver data and rpc it to the COSI controller, which would be at a known network endpoint. 

In either case the sidecar needs to be given a signal that it may begin processing OBs.  This could be a flag on in the ProvisionerRegistration.status or a return value of the rpc.

I realize this is more complicated than CSI's registration.  COSI doesn't have access to the kubelet and isn't bound by node affinity, meaning unix sockets aren't useful as a solution.

1. `phase`: is the current state of the ObjectBucket:
    - `Bound`: the operator finished processing the request and linked the OBC and OB
    - `Released`: the OBC has been deleted, leaving the OB unclaimed.
    - `Failed`: error and all retries have been exhausted.
    - `Retrying`: set when a recoverable driver or kubernetes error is encountered during bucket creation or access granting. Will be retried.

\* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

#### BucketClass

A cluster-wide scoped custom resource.
During greenfield workflows an OBC references a Bucket Class (BC).  The bucket class defines a release policy, and specifies driver specific parameters, such as region, bucket lifecycle policies, etc., as well as the name of the driver as returned by the `GetDriverInfo()` rpc.  The driver name is used to filter OBs meant to be handled by the given driver.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketClass
metadata:
  name: 
  namespace: [1]
driver: [2]  # should this be named "provisioner:" ??

> config: string:string [3]

Reviewer 5: `parameters` would help consistency with StorageClass

Reviewer 3: +1

1. The COSI controller detects the object store bucket and Secret in the BucketClass and creates a BucketContent in the `cosi-system` namespace.
1. The COSI controller generates a Secret in the Bucket's namespace containing the parent Secret's data, with an owner reference to the Bucket.
1. The workflow ingests the Secret to begin operation.

#### Revoke Access

1. The user deletes the Bucket, which blocks until revoke operations complete.
1. The COSI controller deletes the Bucket's bound BucketContent object.
1. The COSI controller removes the finalizer from the Bucket, allowing it and the child Secret to be garbage collected.

##  Custom Resource Definitions

#### Bucket

A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the Bucket is "bound" to the corresponding BucketContent. The is used to prevent further binds to the BucketContent.

```yaml
apiVersion: cosi.io/v1alpha1
kind: Bucket
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
  - cosi.io/finalizer [2]
spec:
  bucketName: [3]
  generateBucketName: [4]
  bucketClassName: [5]
  secretName: [6]

>   protocol: [7]

Author: since it's an array should it be named `protocols`?

Reviewer 3: Is it reasonable to think that an app would be written around more than one client sdk?

> 	accessMode: {"ro", "rw"} [8]

Author: what is the red "tab" ?

Reviewer 3: github calling out a format error.  will fix

1. `accessMode`:  The requested level of access provided to the returned access credentials.
1. `bucketContentName`: Name of a bound BucketContent

> \* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

> ** Ignored in Brownfield and Static operations

#### BucketContent

A cluster-scoped resource representing an object store bucket. The BucketContent is expected to store stateful data relevant to bucket deprovisioning. The BucketContent is bound to the Bucket in a 1:1 mapping.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketContent
Metadata:
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
  - cosi.io/finalizer [3]
spec:
  bucketClassName: [4]
  supportedProtocols: [5]
  releasePolicy: {"Delete", "Retain"} [6]
  bucketRef: [7]
    name:
    namespace:
  secretRef: [8]
  accessMode: {"rw", "ro"} [9]
status:
  bucketAttributes: <map[string]string> [10]
  phase: {"Bound", "Released", "Failed", "Errored"} [11]
  conditions:
```

> 1. `name`: Generated in the pattern of `“bucket-”<BUCKET-NAMESPACE>"-"<BUCKET-NAME>`

Author: what are the k8s truncation rules? What do we do if the `bucket`'s name + ns exceeds 63 chars?

Reviewer 3: Going from memory, the base prefix is truncated by the length of the randomized suffix.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketClass
metadata:
  name: 
provisioner: [1]
supportedProtocols: [2]
accessMode: {"ro", "rw"} [3]
releasePolicy: {"Delete", "Retain"} [4]
bucketContentName: [5]
parameters: string:string [6]
```

1. `provisioner`: Used by sidecars to filter BucketContent objects 
1. `supportedProtocols`: A strings array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.)
1. `accessModes`: Declares the level of access given to credentials provisioned through this class.
1.  `releasePolicy`: Prescribes outcome of a Deletion and Revoke events.
    - `Delete`:  the bucket and its contents are destroyed
    - `Retain`:  the bucket and its contents are preserved, only the user’s access privileges are terminated
- `bucketContentName`: (Optional). An admin defined BucketContent representing a Brownfield or Static bucket.  A non-nil value in this field prevents the BucketClass from being used for Greenfield.

> - `parameters`: object store specific key-value pairs passed to the driver.

Author: This is where a secret name + ns would be defined for object store using a secret to contain the cred needed to create buckets.

Reviewer 3: IMO the parameter list is exclusively used to store driver defined variables.  A discrete field should be included for a secreteRef.

Related - what does this secret represent?  Is it an administrative credential used to affect user policy and bucket creation?  Or is it a user privileged cred meant to be associated with the bucket?  Particularly in brownfield/static, is it already associated with the bucket as a user cred?

The object storage API will also provide a path towards community maintained automatio, which will be capable of bucket lifecycle operations.  It is anticipated that the controller provide an API such that pluggable drivers may be written to implement operations specific to the object store provider.

This proposal does _not_ include a standardized *protocol* or abstraction of storage vendor APIs

## Goals

+ Define a control plane API in order to standardize and formalize Kubernetes object storage representation
+ As MVP, be accessible to the largest groups of consumers by supporting the major object storage protocols (S3, Google Cloud Storage, Azure Blob) while being extensible for future protocol additions.
+ Present similar workflows for both greenfield and brownfield bucket operations.

## Non-Goals

+ Define the _data-plane_ object store interface.
+ Precisely define automation around proposed bucket APIs
+ Manage user identities or roles

##  Vocabulary

> +  _Brownfield Bucket_ - externally created and represented by a [BucketClass](#bucketclass) and managed by a provisioner.

Reviewer 6: I'm a bit confused by terminology here (maybe would become more clear if I continue reading the document).

Basically, when comparing this with existing file/block storage terminology in K8s, a 'brownfield bucket' is like creating a PV backed by an NFS export where this NFS export is somehow already existing, right? However, the goal of a `StorageClass` is not to 'represent' such PV or something along those lines... The PV can still reference its `StorageClass` (the `storageClassName` field) which may impact mount options and such.

Basically, `BucketClass` immediately reminds me of `StorageClass`, but from what I read here, the relation between a brownfield bucket and a `BucketClass` is not alike the relation between a `PersistentVolume` and its `StorageClass`.

Reviewer 3: The overloading of BucketClasses has been addressed in the last couple review meetings.  I've got changes in-flight that propose a different model that we think solves the issue.

+ _Bucket_ - A user-namespaced custom resource representing an object store bucket.

> +  _BucketClass_ - A cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set.

Reviewer 6: This overloading of either "A description of how to provision a new bucket" or "A description of a single existing bucket" seems confusing. Why is the design this way?

### Manually Managed Buckets

Clusters with zero automation around the API still benefit from it when deploying Bucket aware workload-orchistrating controllers.   This case is releveant as it’s likely to be the first step towards adoption for some users.  The app will In this case, admins would define BucketContents themselves and users would create Buckets with a prefilled `bucketContentRef` specifying the BucketContent.  After that, a workload controller would be ready to inject connection and credential data from the BucketContent into the worker pods.

![](/Users/jcope/Workspace/go/src/github.com/yard-turkey/enhancements/keps/sig-storage/bucket-static-api-relation.png)

### Provisioner Secrets

Per [Non-Goals](#non-goals), it is not within the scope of the proposal to abstract IAM operations.  Instead, provisioner and user credentials should be provided to automation by admins or users.  

BucketClass.Parameter key values are used to specify Secret name and namespace.  These may be considered of different degrees of granularity.

> - **Per Provisioner:** the Secret is used for all provisioning operations.  These Secrets should be injected directly in provisioner containers via [common Kubernetes patterns](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/).

Reviewer 6: How would this work if the `Secret` namespace/name are defined in a `BucketClass` `Parameter` field?

Reviewer 3: Do you mean "how do we enforce a heirarchy"? This would ultimately be on admins and driver authors.  We'd have no knowledge of what secrets are mounted to a driver's container.  So, if an admin specifies a secret in the BucketClass, and the driver can accept the credentials via an interface, then it's up to the driver to decide what has priority.  

Admins would have to be aware of driver capabilities in this case.

Reviewer 6: So, a driver should still be able to retrieve `Secret`s from the K8s API?

Let me expand on my first comment: if the `BucketClass` specifies a `Secret` to use (name/namespace), then how could this be injected in the K8s `Pod` where the provisioner is running, unless 'something else' creates this provisioner `Pod(Spec)` to include this `Secret` as a mounted volume, or an environment variable, or whatever? Basically, if the provisioner is already running and should manage multiple `BucketClass`es, how would this work?

Furthermore, exposing a `Secret` as a volume, or environment variables, or whatnot, IIRC only works when the `Pod` sits in the same `Namespace` as the `Secret`, which seems restrictive.

As an example, the following key’s represent a minimun of what automation should support.  `provisioner-secret*` keys specify Secret to be used for bucket operations.  `user-secret-*` keys specify Secrets containing user identity information.

```yaml
cosi.io/provisioner-secret-name:
cosi.io/provisioner-secret-namespace:
cosi.io/user-secret-name:
cosi.io/user-secret-namespace:
```

Bucket automation should support templating of Secret names and namespaces. For example:

```yaml
cosi.io/user-secret-name: "${bucket.name}"
cosi.io/user-secret-namespace: "${bucket.namespace}-region-west"
```

> Note: Annotation key templates are not mentioned as of yet.  However, need for them may arise during automation design.

## Custom Resource Definitions

#### Bucket

> A user facing API object representing an object store bucket. Created by a user in their app's namespace. Once provisiong is complete, the `Bucket` is "bound" to the corresponding `BucketContent`. Binding is 1:1, meaning there is only one `BucketContent` per `Bucket` and vice-versa.

Reviewer 6: Not being able to use a single bucket (in the storage system) from workloads running in  two (or more) namespaces seems quite restrictive...

Reviewer 7: Just like nfs volume(shared filesystem), pvc object is also namespaced.
If their is necessary to share object storage between namespaces, maybe create 2 bucket-bucketcontent pairs which confiured as the same backend.

```yaml
apiVersion: cosi.io/v1alpha1
kind: Bucket
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
  - cosi.io/finalizer [2]
spec:
  protocol: [3]
  bucketPrefix: [4]
  bucketClassName: [5]
  accessSecretName: [6]
status:
  bucketContentName: [7]
  phase: [8]
  conditions: 
```
1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.

> 1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.

Reviewer 6: The `one of` is restrictive, and I'm not sure why there should be such restriction. This value should be opaque to anyone but a provisioner and a workload/client, so no need to restrict its value in the schema.

Reviewer 7: "supportedProtocols" in bucketclass is the protocol list and check this. I am also confused for this, why not support one protocol for one bucketclass? and this restrict is not needed.

Author: From BucketClass description:
```
supportedProtocols: protocols the associated object store supports. Applied when matching Bucket to BucketClasses. Admins may specify all protocols a provisioner supports. Bucket.spec.protocol will be checked against the array prior to provisioning.
```
At one point in the design (and this may still be current @wlan0 ?) the bucketClass reference here would be validated to ensure it supported the requested protocol. If not then hard failure. That's my memory...

Reviewer 8: bucketRequest is set to `InvalidArgument` phase, with a message describing that the bucketClass does not implement the requested protocol.

1. `bucketClassName`: Name of the target `BucketClass`.
1. `userSecretName`: (optional) Secret in the Bucket’s namespace storing credentials to be used by a workload for bucket access.
1. `bucketContentName`: Name of a bound `BucketContent`.
1. `phase`: 
   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations
   - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `BucketContent`.

#### BucketContent

A cluster-scoped resource representing an object store bucket. The `BucketContent` is expected to store stateful data relevant to bucket deprovisioning. The `BucketContent` is bound to the `Bucket` in a 1:1 mapping. For MVP a `BucketContent` is not reused.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketContent
Metadata:
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
  - cosi.io/finalizer [3]
spec:
  provisioner: [4]
  releasePolicy: [5]
  accessMode: [6]
  bucketClassName: [7]
  bucketRef: [8]
    name:
    namespace:
    uuid:
  secretRef: [9]
    name:
    namespace:

>   protocol: [10]

Reviewer 6: Similar to a comment above. I'm not sure this schema should restrict/type the value of this field, and instead would suggest using a `parameters map[string]string` approach.

Reviewer 6: Rationale: a provisioner, even for S3-compatible storage for example, may want to keep more meta-information for whatever reason.

Reviewer 3: There were a couple reasons for this decision (having considered string:string maps early on):

1.  An API that just passes an opaque map to a backend driver is overly generic.  So generic in fact that it no longer adheres to the scope for which it's named.  Drivers could be written to do any number of tasks beyond object storage provisioning through it.  We felt this devalued the API.

2. Strictly defining protocol fields (while providing room for expansions) allows us to deliver connection information in a standard format, rather than an arbitrary format defined by drivers.  The result is that apps won't be coupled to drivers.  IOW, an app moving from an on-prem S3 object store to a cloud S3 object store won't need to be reconfigured to ingest the new BucketContent format, even if the drivers differ.

    type: ""
    azureBlob: [11]
      containerName:
    s3: [12]
      endpoint:
      bucketName:
      region:
      signatureVersion:
    gcs: [13]
      bucketName:
      privateKeyName:
      projectId:
      serviceAccount:
  parameters: [14]
status:
  message: [15]
  phase: [16]
  conditions:
```
1. `name`: Generated in the pattern of `<BUCKET-CLASS-NAME>'-'<RANDOM-SUFFIX>`. 
1. `labels`: should be added by controller..  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

> 1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.

Reviewer 6: If the `Bucket` can't be deleted before the `BucketContent` is, then is the `secretRef` here necessary, since this could be retrieved from the matching `Bucket`?

Author: @NicolasT (Note: in the latest version of the KEP the API name `BucketContent` is replaced by `Bucket`)
The finalizers allow COSI to orchestrate deletion. They typical trigger is the user deletes their `BucketRequest`; however, an admin could delete a `Bucket` and we need to account for that. <-- @wlan0 ?

---
mmrrtitle: Object Bucket Provisioning
authors:
  - "@jeffvance"
  - "@copejon"
owning-sig: "sig-storage"
reviewers:
  - "@saad-ali"
  - "@alarge"
  - "@erinboyd"
  - "@guymguym"
  - "@travisn"
approvers:
  - TBD
editor: TBD
creation-date: 2019-11-25
last-updated: 2020-04-28
status: provisional
---

> # Object Bucket Provisioning

Reviewer 9: There's an error that says "error validating KEP metadata: missing key title".
This should be:
```
# Title

Object Bucket Provisioning
```

Author: @xing-yang I made the changes to title but still get validation err. I think I misunderstand your comment.

  secretName: [6]
  bucketName: [7]
status:
  phase: [9]
  conditions: 
```

1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
1. `finalizers`: should be added by controller to defer BucketRequest deletion until backend deletion ops succeed.
1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
1. `bucketPrefix`: (Optional) prefix prepended to a randomly generated bucket name, eg. “yosemite-photos-". If empty no prefix is prepended.
1. `bucketClassName`: Name of the target BucketClass.
1. `secretName`: (optional) Secret in the BucketRequest's namespace storing credentials to be used by a workload for bucket access.
1. `bucketName`: Name of the Bucket instance representing the desired storage instance endpoint.  Set by the COSI Controller for new Buckets.  Specified by the user when connecting to existing buckets.
1. `phase`: 
   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations

>    - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.

Reviewer 3: Good question.  We had planned on this being expanded as we prototype the automation.  We felt there's likely to be some intermittent phases that are worth communicating.

>    - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.

Author: When the delete case is considered and we are thinking of async deletes, we likely need a "DeletePending" phase. There has been some discussion about various "pending" phases, eg. "CreatePending", "GrantPending", "RevokePending". These phases would also be seen in the `Bucket` resource. The idea is that the admin and user can more easily see the current step should something get stuck...

>    - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.

Reviewer 8: In case of errors, or stuck states, we should following the way of the pod statuses. Have a error status

- ErrBucketClassDoesNotSupportProtocol
- ErrBucketDeletionInProgress
etc. 

To answer @xing-yang's question. It might be better to keep them distinct states rather than use a boolean. it'll allow us to expand in the future to add more states. In case of the boolean, we will have a harder time adding new states without adding a new field, or breaking existing API.

#### Bucket

A cluster-scoped resource representing a storage instance.  A Bucket should contain enough information to enable the requesting client to access the storage instance.  At a minimum, a Bucket must store enough identifying information such in order for drivers to accurately target the storage instance (e.g. during a deletion process).

```yaml
apiVersion: cosi.io/v1alpha1
kind: Bucket
Metadata:
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
  - cosi.io/finalizer [3]
spec:

>   provisioner: [4]

Reviewer 3: Yes.  For dynamically generated buckets, these would be set by the controller.  That said, in human created Buckets, this isn't guarantee-able without automation to patch it in.  The purpose of the label is strictly a nicety for admins, so maybe it's something we'd want to punt on for MVP.

1. `keySecretName`: name of the *provisioner* generated `Secret` containing access credentials. This `Secret` exists in the provisioner’s namespace and must be copied to the app namespace by the COSI controller.
1. `provisioner`:  name of the provisioner that should handle this `BucketAccess` instance.  Copied from the `BucketAccessClass`.
1. `parameters`:  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.

#### BucketAccessClass

A cluster-scoped API providing a way for admins to specify policies that may be used to access buckets.  These are always applicable in Greenfield, where access is dynamically granted, and only sometimes applicable in Brownfield, where a user identity may already exist, but not yet have access to a bucket.  In this case, a BucketAccessRequest will still specify the BucketAccessClass in order to determine which actions it is granted to a defined bucket.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketAccessClass
metadata: 
  name:
provisioner: [1]
supportedProtocols: [2]
policyActions: [3]
  allow:
  - "*"
  deny:
  - "*"
parameters: [4]

> ```

Reviewer 10: @jeffvance @copejon what about scope of quota(size, no of objects) here?

Author: As with the[ bucket provisioning lib](https://github.com/kube-object-storage/lib-bucket-provisioner), the `parameters` section can be used to pass on quota information. It's unclear to me how quota will work within  object stores existing outside of k8s but referenced by resources in a k8s cluster. 
For now, we can limit the number of instances of these bucket-related CRs, eg. limit the number of `BucketRequest`s.

---
title: Object Bucket Provisioning
authors:
  - "@jeffvance"
  - "@copejon"
owning-sig: "sig-storage"
reviewers:
  - "@saad-ali"
  - "@alarge"
  - "@erinboyd"
  - "@guymguym"
  - "@travisn"
approvers:
  - TBD
editor: TBD
creation-date: 2019-11-25
last-updated: 2020-07-08
status: provisional
---

# Object Bucket Provisioning

## Table of Contents

<!-- toc -->
- [Summary](#summary)
  - [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
  - [Vocabulary](#vocabulary)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
      - [Admin](#admin)
      - [User](#user)
  - [APIs](#apis)
    - [Storage APIs](#storage-apis)
      - [BucketRequest](#bucketrequest)
      - [Bucket](#bucket)
      - [BucketClass](#bucketclass)
    - [Access APIs](#access-apis)
      - [BucketAccessRequest](#bucketaccessrequest)
      - [BucketAccess](#bucketaccess)
      - [BucketAccessClass](#bucketaccessclass)

>   - [Provisioner Secrets](#provisioner-secrets)

Author: After API added section on typical create and delete workflows for green- and brownfield.

Author: Add TOC for gRPC spec

> <!-- /toc -->

Author: add gRPC spec and add timeline section

Reviewer 8: done 👍

Reviewer 11: resolve this? there's quite a few unresolved comments which are irrelevant...

Author: Is your comment related to a "timeline section"? The toc seems to verify now.

# Summary

> This proposal introduces the *Container Object Storage Interface* (COSI), a system composed of Custom Resource Definitions (CRDs), a controller architecture, and a gRPC specification, for the purpose of standardizing object storage representations in Kubernetes.  Goals and non-goals set the scope for the proposal by defining higher level objectives.  The vocabulary section defines terminology.  User stories illustrate how these APIs may fulfill cluster user requirements.  Relationships between the APIs are provided to illustrate the interconnections between object storage APIs, users' workloads, and object store service instances.  Lastly, the documents states the proposed API specs for the Bucket, Bucket, and BucketClass objects.

Author: "Lastly, the documents states the proposed API specs for the BucketRequest, Bucket,  BucketClass, etc. objects."

While absolute portability cannot be guaranteed because of incompatibilities between providers, workloads reliant on a given protocol (e.g. one of S3, GCS, Azure Blob) may be defined in a single manifest and deployed wherever that protocol is supported.

This proposal does _not_ include a standardized *protocol* or abstraction of storage vendor APIs

## Goals

+ Specify object storage Kubernetes APIs for the purpose of orchestrating object store operations
+ Implement a Kubernetes controller architecture with support for pluggable provisioners
+ As MVP, be accessible to the largest groups of consumers by supporting the major object storage protocols (S3, Google Cloud Storage, Azure Blob) while being extensible for future protocol additions.
+ Present similar workflows for both greenfield and brownfield bucket operations.

## Non-Goals

+ Define the _data-plane_ object store interface to replace or supplement existing vendor interfaces (i.e. replace GCS, S3, or Azure Blob)

##  Vocabulary

> +  _Brownfield Bucket_ - a storage instance created out of band.

Author: "an existing storage instance which could be created within or external to a Kubernetes cluster."

> +  _Brownfield Bucket_ - a storage instance created out of band.

Author: add:
  _adapter_ - a pod which receives Kubelet gRPC `nodePublish` and `nodeUnpublish` requests, ensure the target bucket has been provisioned, and notifies the kubelet that the pod can be run.

+ _BucketRequest_ - A user-namespaced custom resource representing a request for a storage instance endpoint.
+  _BucketClass_ - A cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set for creating new buckets

> + _Bucket_ - A cluster-scoped custom resource referenced by a Bucket and containing connection information and metadata for a storage instance.

Author: "...referenced by a `BucketRequest`..."

> + _Greenfield Bucket_ - a bucket created by automation.

Author: a new bucket created...

+  _Object_ - An atomic, immutable unit of data stored in buckets.
+  *Storage Instance* - Refers to the back object storage endpoint being abstracted by the Bucket API (a.k.a “bucket” or “container”).

> + _Driverless_ - A system where not driver is deployed to automate object store operations.  COSI automation may still be deployed to managed COSI APIs.  See [Architecture](#architecture).

Author: "A system where no driver is ..."

Author: Driverless was used as a synonym for "static brownfield" and there is a reference to static provisioning in the  BucketClass.provisioner description. But this concept is not mentioned anywhere else in the KEP. Maybe we are not longer supporting static/driverless for MVP??  @wlan0

# Proposal

## User Stories

#### Admin

> - As a cluster administrator, I can manage multiple object store providers via the Kubernetes API, so that I can see all object storage instances from a single pane, regardless of the backing storage vendor.

Author: "manage" is vague.  What about:
- As a single cluster administrator, I can enable multiple object store providers such that users (app developers) can create new, or be granted access to existing buckets, all via the Kubernetes API.

- As a cluster administrator, I can see all object storage instances from a single pane, regardless of the backing storage vendor.

- As a cluster administrator, I can control access granted to new and existing buckets when accessed from the cluster.

Reviewer 8: > As a cluster administrator, I can control access granted to new and existing buckets when accessed from the cluster.

... , regardless of the backing object storage vendor

#### User

- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.

> - As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to object storage instances.

Author: Is an example of this Crossplane?

## APIs

> NOTE: For a graphical mapping of how these APIs relate to each other, see [API Relationships](#api-relationships).

### Storage APIs

#### BucketRequest

A namespaced API representing a workload’s need for a storage instance endpoint. Created in the workload's namespace.  A [BucketAccessRequest](#bucketaccessrequest) is required in order to gain credentialed access to the storage instance.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketRequest
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
  - cosi.io/finalizer [2]
spec:
  protocol: [3]
  bucketPrefix: [4]

>   bucketClassName: [5]

Author: bucketclass is only used for greenfield - optional.

Reviewer 11: @jeffvance This comment is conflicting with the [5] explanation later - which one is more uptodate?

Author: We've gone back and forth several times on requiring BC for brownfield. Good catch.

  secretName: [6]
  bucketName: [7]
status:
  phase: [9]
  conditions: 
```

> 1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

Author: "labels : added by the controller. Key's..."

1. `finalizers`: should be added by controller to defer BucketRequest deletion until backend deletion ops succeed.
1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
1. `bucketPrefix`: (Optional) prefix prepended to a randomly generated bucket name, eg. “yosemite-photos-". If empty no prefix is prepended.

> 1. `bucketClassName`: Name of the target BucketClass.

Author: add (if this is correct!) that the bucket class's supportedProtocols list is checked against the requested protocol and if no match then and error is logged and the bucket request is ignored. Probably status should be updated and an event recorded so that it's easier for the user to know what's happened.

1. `secretName`: (optional) Secret in the BucketRequest's namespace storing credentials to be used by a workload for bucket access.
1. `bucketName`: Name of the Bucket instance representing the desired storage instance endpoint.  Set by the COSI Controller for new Buckets.  Specified by the user when connecting to existing buckets.
1. `phase`: 
   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations

>    - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.

Author: "Bound: Provisioning operations have completed and the BucketRequest has been bound to a Bucket."

#### Bucket

> A cluster-scoped resource representing a storage instance.  A Bucket should contain enough information to enable the requesting client to access the storage instance.  At a minimum, a Bucket must store enough identifying information such in order for drivers to accurately target the storage instance (e.g. during a deletion process).

Author: "...enough identifying information in order for drivers"

> A cluster-scoped resource representing a storage instance.  A Bucket should contain enough information to enable the requesting client to access the storage instance.  At a minimum, a Bucket must store enough identifying information such in order for drivers to accurately target the storage instance (e.g. during a deletion process).

Author: Mention mapping of `BucketRequest` to `Bucket`. Unlike PVC <-> PV (1:1 mapping) there can be 1 `Bucket` for many `BucketRequest`s.

```yaml
apiVersion: cosi.io/v1alpha1
kind: Bucket
Metadata:
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
  - cosi.io/finalizer [3]
spec:
  provisioner: [4]
  releasePolicy: [5]
  anonymousAccessMode: [6]
    - private
    - publicRead
    - publicReadWrite
  bucketClassName: [7]
  permittedNamespaces: [8]
    - name:
      uid:
  protocol: [9]
    protocolSignature: ""
    azureBlob: [10]
      containerName:
      storageAccount:
    s3: [11]
      endpoint:
      bucketName:
      region:
      signatureVersion:
    gcs: [12]
      bucketName:
      privateKeyName:
      projectId:
      serviceAccount:
  parameters: [13]
status:
  message: [14]
  phase: [15]
  conditions:
```

> 1. `name`: For Greenfield, generated in the pattern of `bucket-<my-bucket>-<my-bucket-namespace>`. 

Author: name format should be more explicit: _bucket-<bucketRequest.ame>-<bucketRequest.namespace>_

> 1. `labels`: should be added by controller..  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

Author: delete "should be" here and finalizer

1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
1. `provisioner`: The provisioner field defined in the `BucketClass`.  Used by sidecars to filter Buckets.
1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield cases, backing storage instance is Retained, and never deleted.
   - _Delete_:  the bucket and its contents are destroyed.
   - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed.

> 1. `anonymousAccessMode`:  ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.  If `BucketClass.anonymousAccessMode` is set, the value is copied here.  Accepted values:

Author: slight wordsmithing...

Author: Need to be clear if these values pertain to the objects within the bucket, or only to the bucket itself. Eg. R can mean user can read objects contained in bucket, or it could mean user can read the bucket (eg. to get a list of its objects).

Reviewer 8: Good question. There is nuanced access modes in object storage systems. ListBucket and getObjectis a different operations, and hence require different permissions. In the case of anonymousAccessMode, we are talking about a more general access pattern, similar to what is seen in file systems. i.e. In case of anonymousAccessMode 'R_ONLY', we allow both ListBucket and GetObject. In case of 'RW', we allow ListBucket,GetObject and PutObject and in case of W_ONLY, we only allow putObject.

   - `private`: Default, disallow uncredentialed access to the storage instance.
   - `ro`: Read only, uncredentialed users are permitted *read* operations for objects within the storage instance.
   - `rw`: Read/Write, uncredentialed users are permitted *read and write* operations for objects within the storage instance. 
   - `wo`: Write, uncredentialed users are permitted *write* operations for objects within the storage instance. 
1. `bucketClassName`: Name of the associated BucketClass.
1. `permittedNamespaces`: An array of namespaces, identified by a name and uid, from which  BucketRequests are allowed to bind to the Bucket.  Provided to allow admins a layer of cluster-layer access control.  Does **not** reflect or alter the backing storage instances' ACLs or IAM policies.
   - In Greenfield, the originating BuckerRequest’s namespace must be specified at time of Bucket generation.
   - In Brownfield, this list is defined by the admin.
1. `protocol`: The protocol the application will use to access the storage instance.

>    - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:

Author: what is the difference between _protoco_ and _protocolSignature_? @wlan0

     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
     - `s3`: data required to target a provisioned S3 bucket and/or user.
     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
1. `parameters`: a copy of the BucketClass parameters.
1. `message`: a human readable description detailing the reason for the current `phase``.

> 1. `phase`: is the current state of the Bucket:

Author: probably we need some _xyzPending_ phases added.

Reviewer 8: Need to have a longer discussion about this. Lets talk about it in tomorrow's meeting.

Reviewer 8: Created
Bound
Released
ErrXYZ

   - _Ready_: the controller finished processing and the controller is ready to receive access requests.
   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
   - _Failed_: error and all retries have been exhausted.
   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.

#### BucketClass

> 

Author: Is BucketClass _only_ used in greenfield?  @wlan0

Reviewer 8: BucketClass is also used in brownfield requests. It is used to retrieve a bucket that satisfies the constraints laid out in the bucketClass

Author: Per conversation with @wlan0 : BC is only used for greenfield

Reviewer 11: @jeffvance @wlan0  If BC is for greenfield only then for brownfield we still need `B.spec.allowedNamespaces` so admin who creates it can define which namespaces can bind to it. 

The only "confusion" we might want to clarify is in the greenfield case when BC.allowedNamespaces=[X,Y,Z] and namespace X creates a greenfield BR with this BC, then COSI provisions a new bucket, creates a new B, bind it to the BR, etc. But what does `B.allowedNamespaces` mean for greenfield Buckets? Is it a copy of the `BC.allowedNamespaces=[X,Y,Z]`? or is it restricted just to the namespace that provisioned it - B.allowedNamespaces=[X]? I think that copying from the BC would lead to a confusion - could namespace Y accidently/deliberately bind to namespace X greenfield bucket?

Author: > what does B.allowedNamespaces mean for greenfield Buckets? Is it a copy of the BC.allowedNamespaces=[X,Y,Z]?

It is a copy from the BC. The list can be mutated by the admin.

Reviewer 11: @jeffvance I still feel like for greenfield it doesn't make any sense to copy allowedNamespaces from the BC into the B.

It's as if by sharing a BC between different users I am somehow allowing other namespaces to access my data.

Even if you somehow make sure the other namespaces cannot bind to this B, even declaratively this sounds like a bad idea to me. Is that concerning you as well?

Author: For greenfield, `BC.allowedNamespaces` gives the admin some control over who can create new buckets. Once a B is created based on a BR and the BC and bound to the BR, I agree that B.allowedNamespaces is not needed by COSI due to our 1:1 mapping of BR:B. And, B.allowedNamespaces could be set to the BR's ns and changed for brownfield after the admin clones this B.

For pure brownfield (not green -> brown), the admin creates a `Bucket` instance and defines which namespaces are allowed to request access to the B.  Here, the user BR does not need to reference a BC since the B already exists. But the KEP has gone back and forth on this. B.allowedNamespaces gives the admin control over who can request access to the B. Eg. a cluster scoped RBAC could allow all users to see Bs  but admin still wants to control who can use a B.

For the green->brownfield case, if B.allowedNamespaces contains only the ns of the BR which caused this B to be instantiated, then the admin needs to change this field after the copy.  This can work fine since the admin might also need to edit B.allowedNamespaces even if this value is based on BC.allowedNamespaces.

> A cluster-scoped API whose purpose it to provide admins control over the handling of bucket provisioning by specifying provisioner-defined key-values in the `parameters` map.  The `BucketClass` defines a release policy, and specifies driver specific parameters and the provisioner name.

Author: add greenfield only

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketClass
metadata:
  name: 
provisioner: [1]
isDefaultBucketClass: [2]
supportedProtocols: {"azureblob", "gcs", "s3", ... } [3]
anonymousAccessMode: {"ro", "wo", "rw"} [4]
additionalPermittedNamespaces: [5]
- name:
  uid: 
releasePolicy: {"Delete", "Retain"} [6]
parameters: [7]
```

> 1. `provisioner`: The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.

Author: What _secretRef_ does this refer to?
And, is static provisioning supported?  
@wlan0

Reviewer 8: UUIC, staticProvisioning is where buckets are created manually with specified names rather than where Kubernetes auto-generates a name and creates it on the fly.

> 1. `isDefaultBucketClass`: boolean. When true, signals that the controller should attempt to match `Bucket`’s without a defined `BucketClass` to this class, accounting for the `Bucket`’s requested protocol.  Multiple default classes for the same protocol will produce undefined behaviour.

Author: Add: Multiple default bucket classes are allowed as along as their protocols differ. Two default bucket classes with the same protocol results in undefined behavior and should be avoided.

1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses. Admins may specify all protocols a provisioner supports.  `Bucket.spec.protocol` will be checked against the array prior to provisioning.
1. `anonymousAccessMode`: (Optional) ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.

> 1. `additionalPermittedNamespaces`: A list of namespace *in addition to the originating namespace* that will be allowed access to this Bucket.  A nil and empty string value are equivalent and must not append any namespace to `Bucket.spec.permittedNamespaces` list.

Author: nil ("") sentence can be deleted...

> 1. `additionalPermittedNamespaces`: A list of namespace *in addition to the originating namespace* that will be allowed access to this Bucket.  A nil and empty string value are equivalent and must not append any namespace to `Bucket.spec.permittedNamespaces` list.

Author: @wlan0 Is this field needed to handle the _greenfield_ -> _brownfield_ shared use-case?

Author: _Maybe_ we should delete this field until we can provide solid justification. Easier to add a field than to remove (I think)...

> 1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield and Static cases, *Retain* is mandated. 

Author: do we still have the concept of static?

Reviewer 8: Yeah, the above statement is true.

>    - `Delete`:  the bucket and its contents are destroyed

Author: @wlan0 will COSI delete contents in _any_ situation?

Reviewer 8: Yes, if release policy is delete

   - `Retain`:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed

> 1. `parameters`:  (Optional)  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.

Author: Add note to see special Secret key names in the last section of the KEP.

### Access APIs

> The Access APIs abstract the backend policy system.  Access policy and user identities are an integral part of most object stores.  As such, a system must be implemented to manage both user/credential creation and the binding of those users to individual buckets via policies.  Object stores differ from file and block storage in how they manage users, with cloud providers typically integrating with an IAM platform.  This API includes support for cloud platform identity integration with Kubernetes ServiceAccounts.  On-prem solutions usually provide their own user management systems, which may look very different from each other and from IAM platforms.  We must also account for third party authentication solutions that may be integrated with an on-prem service.

Author: "...Access APIs abstracts the backend..."

#### BucketAccessRequest

> A user facing API representing an object store user and an access policy defining the user’s relation to a storage instance.  A user will create a BucketAccessRequest in the app namespace.  A BucketAccessRequest can specify *either* a ServiceAccount in the same namespace or a desired Secret name.  Specifying a ServiceAccount enables provisioners to support cloud provider identity integration with their respective Kubernetes cluster offerings.

Author: mention this is also the same ns of the `BucketRequest`.

Author: @wlan0 what is the mapping between `BucketRequest`s and `BucketAcccessRequest`s? Is it 1:1, 1:many, many:1 or m:n?

Reviewer 8: Yes, for each bucketrequest, we get credentials back from the backend. The input to the backend will be captured in bucketAccessRequest

Author: "A BucketAccessRequest can specify *either* a ServiceAccount in the same namespace or a desired Secret name."
I do not see a secret name anywhere in the BAR.

Reviewer 8: It should be BucketRequest, which is what holds the desired secretName

> 

Author: what happens if the user deletes the BAR before (or instead of) the BR?  @wlan0 ?

Reviewer 8: the bucketRequest is not dependent on the bucketAccessRequest. The bucket creation will take place as usual.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketAccessRequest
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
  - cosi.io/finalizer [2]
spec:
  serviceAccountName: [3]

>   bucketRequest: [4] 

Author: A BucketAccessRequest references a BucketRequest which may or may not need to exist? But this is a one-way link since a BucketRequest does _not_ reference a BucketAccessRequest.

Reviewer 8: Lets discuss this further tomorrow

Author: Update: discussing with @wlan0 : the field name should be `bucket` not bucketRequest

Author: @wlan0 the field wil be named `bucket`, per above. Is this field omitted for greenfield and required for brownfield (like in provisioning)? I assume COSI will fill in this name in greenfield. In brownfield the admin, who has to create the `Bucket` will also manually create the BA and the user needs to know the name of the BA when she creates her BAR. Is this correct?

Reviewer 8: It is possible to set it for the greenfield case as well, say when requesting to create a bucket by a certain name.

>  In brownfield the admin, who has to create the Bucket will also manually create the BA, and the user needs to know the name of the BA when she creates her BAR. Is this correct?

The credentials can be auto-provisioned even in the case of brownfield. This can be signaled by leaving the bucketAccessName empty.

  bucketAccessClassName: [5]
  bucketAccessName: [6]
status:
  message: [7]
  phase: [8]
```

1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
1. `finalizers`: should be added by controller to defer `BucketAccessRequest` deletion until backend deletion ops succeed.

> 1. `serviceAccountName`: the name of a Kubernetes ServiceAccount in the same namespace.  This field is included to support cloud provider identity integration.  Should not be set when specifying `accessSecretName`.

Author: I don't see the mentioned _accessSecretName_

Author: @wlan0 answer: `bucketRequest.secretName`

Reviewer 8: This should be updated. The secretName field is missing from the BAR object. That is the field this above "accessSecretName" is referring to.

> 1. `bucket`: The the name of the bucket to which the user identity or ServiceAccount should be granted access to, according to the policies defined in the `BucketAccessClass`.

Author: 1. `bucketRequest`: The of the BucketRequest (in this namespace) to which ...

Author: @wlan0 does the BucketRequest need to exist before creation of a BucketAccessRequest?

Reviewer 8: BucketAccessRequest is the input to the cosi-controller-manager for creating BucketAccess

> 1. `bucketAccessClassName`: name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.

Author: Is this required, or is there a default BAC, like we have for BCs?  @wlan0 
Note: there is no default BAC described in the KEP.

Reviewer 8: I believe we should have one. It should depict the default access mode for bucket consumers.

> 1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance

Author: @wlan0 is this filled in by COSI for greenfield? Is there a brownfield case where the user fills this field in?

Reviewer 8: yes, in both cases it is filled in by COSI

1. `message`: a human readable description detailing the reason for the current `phase``.
1. `phase`: is the current state of the Bucket:
   - _Bound_: the controller finished processing and has bound the BucketAccess to the BucketRequest via the BucketRequest.Spec.BucketAccessRequestName and BucketRequest.Spec.BucketAccessRequestNamespace

>    - _Failed_: error and all retries have been exhausted.

Author: When are _all retries exhausted_? I thought, in general, k8s will continue to retry and this is reflected in controller-runtime.

Reviewer 8: Yeah, the states should be 

Created
Bound

and various
ErrXYZ states

   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.

#### BucketAccess

> A cluster-scoped administrative API which encapsulates fields from the BucketAccessRequest and the BucketAccessClass.  The purpose of the API is to server as communication path between provisioner and a central COSI controller.  In Greenfield, the COSI controller will create BucketAccess instances for new BucketAccessRequests.  The relationship between a BucketAccessRequests and a BucketAccess should be 1:1.  

Author: "...purpose of the API is to serve as..."

Author: The diagram shows a `BucketAccess` and `Bucket` referencing _each_ other but I don't see those fields in either manifest.

Reviewer 8: the reference will be indicated in the GRPC spec. A bucketRequest and its corresponding bucketAccessRequest are needed to create the bucketAccessObject and the bucket Object

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketAccess
metadata: 
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
  - cosi.io/finalizer [3]
 spec:
  bucketAccessRequestName: [4]
  bucketAccessRequestNamespace: [4]
  serviceAccountName: [5]
  accessSecretName: [6]
  provisioner: [7]
  parameters: [8]
 status:
  message: [9]
  phase: [10]
```

> 1. `name`: For Greenfield, generated in the pattern of `bucketAccessRequest-<my-bucketAccessRequest>-<my-bucketAccessRequest-namespace>`. 

Author: name format should be more explicit:
"bucketAccess-"<bucketAccessRequest.name>"-"<bucketAccessRequest.namespace> (or maybe reverse ns and name?)
However, as is true for many k8s names of this pattern, it is ambiguous where the name ends and the namespace begins...

Author: Should the `BucketAccess` have a protocol struct similar to what we have for a `Bucket`?  @wlan0

> 1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

Author: remove "should be here and below"

1. `finalizers`: should be added by controller to defer `BucketAccess` deletion until backend deletion ops succeed.
1. `bucketAccessRequestName`/`bucketAccessRequestNamespace`: name and namespace of the bound `BucketAccessRequest`
1. `serviceAccountName`: name of the Kubernetes ServiceAccount specified by the `BucketAccessRequest`.  Undefined when the `BucketAccessRequest.accessSecretName` is defined.

> 1. `  accessSecretName`: name of the *provisioner* generated `Secret` containing access credentials. This `Secret` exists in the provisioner’s namespace and must be copied to the app namespace by the COSI controller.

Author: mention that the provisioner's ns is derived from the `provisioner` value.

> 1. `provisioner`:  name of the provisioner that should handle this `BucketAccess` instance.  Copied from the `BucketAccessClass`.

Author: Are we allowing separate provisioners for creating/provisioning buckets vs. granting access to buckets? IOW, 1) can a single provisioner do both (and thus provisioner name here in in the bucket class are the same), or 2) are 2 separate provisioner _required_, or 3) does COSI make no requirement here?

Reviewer 8: the provisioner for creating/deleteing buckets will be the one granting and revoking access as well. They go hand in hand.

> 1. `parameters`:  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.

Author: @wlan0 is this different from `bucketAccessClass.parameters`?

Reviewer 8: This is an additonal set of parameters on top of what is available in the class. An example use case would be where the bucketClass has a set of defaults and this field allows further customization on top of that.

Author: > where the bucketClass has a set of defaults and this field allows further customization on top of that.

@wlan0 1) do you mean BucketClass or BucketAccessClass above?
2) I do not understand if there are any differences between green- and brownfield for BARs, BACs, and BAs like there are with bucket provisioning.
3) I assume that a BA is created manually by the admin who also creates the BAC.

Reviewer 8: > do you mean BucketClass or BucketAccessClass above? 

yes, it should've been bucketAccessClass above

> 3)  I assume that a BA is created manually by the admin who also creates the BAC.

This is one possibility. It is also possible that the BA for brownfield is autogenerated, for instance, in the Greenfield -> brownfield case, each bucket accessRequest gets a new set of credentials for that workload

1. `message`: a human readable description detailing the reason for the current `phase``.
1. `phase`: is the current state of the Bucket:
   - _Bound_: the controller finished processing the request and has bound the BucketAccess to the BucketAccessRequest
   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
   - _Failed_: error and all retries have been exhausted.
   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.

#### BucketAccessClass

A cluster-scoped API providing a way for admins to specify policies that may be used to access buckets.  These are always applicable in Greenfield, where access is dynamically granted, and only sometimes applicable in Brownfield, where a user identity may already exist, but not yet have access to a bucket.  In this case, a BucketAccessRequest will still specify the BucketAccessClass in order to determine which actions it is granted to a defined bucket.

> 

Author: Is there a default BAC @wlan0 ? If yes is there at most one default BAC, or does protocol somehow fit in such there there could be >1 default BAC?

Reviewer 8: I think there should be one BAC per protocol, similar to BC

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketAccessClass
metadata: 
  name:
provisioner: [1]
supportedProtocols: [2]

> policyActions: [3]

Author: I believe this is the only field unique to a bucket access class when comparing to a bucket class. 
@saad-ali @wlan0 do we _really_ need a bucket access class? I understand separation of concerns, but does it really simplify the use cases by much?

  allow:
  - "*"
  deny:
  - "*"
parameters: [4]
```

1. `provisioner`: The name of the provisioner that `BucketAccess` instances should be managed by.

> 1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses.  Admins may specify more than protocol when necessary.  `BucketAccessRequests.spec.protocol` will be checked against this array prior to provisioning.

Author: "Applied when matching BucketAccessRequests to BucketAccessClasses."
"Admins may specify more than one protocol when necessary. "

Author: "Applied when matching BucketAccessRequests to BucketAccessClasses."
"Admins may specify more than one protocol when necessary. "

Reviewer 12: Multiple default classes are allowed, as long as they are for different protocols. At the same time, a single class can specify multiple protocols. Isn't that confusing? Wouldn't it be easier to have a 1:1 mapping between class and protocol? Additional parameters will likely be protocol-dependent as well, so I'm not sure if mixing multiple protocols in the same class is a good idea.

Author: @wonderland GCP and Azure support >1 protocol. The idea is that the BC represents capabilities of the underlying object store and a BR (bucket request) knows which protocol to request. COSI will validate that the referenced BC supports the desired protocol.
Regarding multiple default BCs: our thinking was that an admin could define one default BC per protocol and COSI will find the correct default BC by matching BR's protocol with supported protocols in all default BCs (of which there should not be more than 3). If default BC-1 supports "s3" and default BC-2 supports "gcp, s3" then we have undefined behavior when a BR requests s3. 
We could state that default BCs can only list one supported protocol to make it less confusing. Thoughts?

Reviewer 12: Many on-premises solutions also support several protocols (S3, Swift, ...), but each protocol would typically come with slightly different parameters and capabilities. So mixing them in a single BC, possibly making exceptions to that rule for default BC, seems confusing to me. I would find it more logical to create (at least) one BC for each protocol the underlying object store supports (if you want to use all protocols). In most cases it will probably be just a single protocol anyway, so this wouldn't lead to an inflation of BC on a cluster. But maybe that's just me getting confused easily ;-)

Author: @wonderland I agree it's probably less confusing for a default BC to support only 1 protocol, especially when considering _parameters_.  I am fine with this. @wlan0 ?

Author: @wonderland this discussion about default classes should be under `BucketClass` not `BucketAccessClass`.

1. `policyActions`: a set of provisioner/platform defined policy actions to allow or deny a given user identity.
1. `parameters`:   (Optional)  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.

---

## COSI Architecture (Warning - Construction Area)

### Topology

![](./bucket-api-diagrams-topology.png)

### Greenfield

In an automated system where bucket lifecycles are managed by a controller:

1. A BucketClass is defined to enable the provisioning of object store storage, with a specified configuration.

> 2. A user will define a [Bucket](#bucket) with the [BucketClass](#bucketclass) specified.  

Author: "... define a BucketRequest..."

> 3. A [Bucket](#Bucket) is instantiated and encapsulates all configuration information from the BucketClass and Bucket.  The Bucket will store all necessary information about the provisioned storage instance, including connection data.  Credentials and other sensitive information should not be stored in a Bucket. 

Author: "... information from the BucketClass and BucketRequest."

4. A driver will then provision the object storage and return the connection information, to be written to the Bucket.

#### Sharing Dynamically Created Buckets

Once a [Bucket](#Bucket) is created, it is discoverable by other users in the cluster.  

> 5. In order to access the Bucket, a user may create a [Bucket](#bucket) that specifies the Bucket by name.

Author: "...a user may create a [BucketRequest](#bucketrequest) that specifies the Bucket by name."

> > Note: Per [Non-Goals](#non-goals), access management is not within the scope of this KEP.  ACLs, access policies, and credentialing should be handled out of band.

Author: This is _not_ stated in the non-goals section. Should it be? @wlan0 ?

Reviewer 8: Yes, we should state them in Non-Goals. Good catch

### Brownfield

> “Brownfield” describes any case where the bucket already exists in the cloud service and is *not yet represented* by a [Bucket](#Bucket) instance.  In order to publish the cloud service bucket to the cluster:

Author: "... and is *not yet represented* by a [Bucket](#Bucket) instance"
@wlan0 I think this is not correct. Multiple BRs can reference the same BC and the same Bucket.

Reviewer 8: This is not entirely true anymore right? Yeah, second reference to a bucket created by COSI is allowed. We need to fix this and say that brownfield describes any case where the bucket already exists in the backend storage.

1. The bucket should already be provisioned in the cloud service.
1. A Bucket is created to represent the provisioned bucket.  

> 1. Once done, Bucket API instances may be created by users, specifying the desired Bucket.

Author: "Once done, BucketRequest API instances..."

> >  Note: Per [Non-Goals](#non-goals), access management is not within the scope of this KEP.  ACLs, access policies, and credentialing should be handled out of band.  Buckets may be configured to whitelist access for specific namespaces.

Author: s/whitelist/allow/

- **Per Provisioner:** the Secret is used for all provisioning operations, for the lifetime of the provisioner.  These Secrets should be injected directly into the provisioner's container via [common Kubernetes patterns](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/).

Credentials may also be specified at more granular levels in order to allow for context dependent keys.  E.g. When authorization differs between BucketClasses or between individual operations.  This may be facilitated by defining a set of string keys which the core automation will be aware of, so that Secrets may be referenced by BucketClasses.  For example:

```yaml
cosi.io/provisioner-secret-name:
cosi.io/provisioner-secret-namespace:
```

- **Per BucketClass:** A secret may be made specific to a BucketClass.  This suits cases where authorization may be segregated in the object store.  The Secret may then be defined explicitly in the `bucketClass.parameters` map.

  ```yaml
  cosi.io/provisioner-secret-name: "foo"
  cosi.io/provisioner-secret-namespace: "bar"
  ```

> - **Per Operation/Bucket:** Unique credentials are passed per Bucket or operation. In order to support dynamic Secret naming, templating similar to [CSI Secret templating](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html) may be used.  E.g.

Author: If this is at all difficult or risky to implement I suggest saying that it is not a part of MVP.

Reviewer 8: That is correct. This cannot be fulfilled as a part of the MVP

### Goals

+ Relieve object store providers from Kubernetes controller details when all they want is to provision bucket storage.
+ Minimize technical ramp-up for storage vendors who have already contributed external storage provisioning (file and block).
+ Make bucket consumption similar to file or block consumption using familiar concepts and commnands.
+ Use native Kubernetes resources where possible, again, to keep the bucket experience for users and admins similar to existing storage provisioning.
+ Be unopinionated about the underlying object-store and at the same time provide a flexible API such that provisioner specific features can be supported.
+ Ensure the bucket consuming pod waits until the target bucket has been created and is accessible. Thus there is no specific order required for when a bucket claim is created vs. when the app pod is run.
+ Present similar user and admin experiences for both _greenfield_ (new) and _brownfield_ (existing) bucket provisioning.

### Non-Goals

+ Update the native Kubernetes PVC-PV API to support object buckets.

> + Support and manage object store data planes.

Reviewer 13: I don't like the POSIX comparison. POSIX has ACLs, write/read verbs, etc.

read/write verbs are not in scope, per above.

> + Support and manage object store data planes.

Reviewer 6: Agree with @copejon. It's not a *comparison*, it's making very clear this is a *non-goal*. From experience, way too many people conflate object storage with file-like access. Making clear this is *not* a project to provide any kind of POSIX/file-like access to object storage would definitely be a clarification for some.

+ an _ObjectBucket_ (OB), similar to a PV, is the k8s representation of the provisioned bucket and may contain object store-specific data needed to allow the bucket to be de-provisioned (e.g. deleted). The OB resides in the same namespace as the object store, and is typically not seen by bucket consumers.
**Note:** namespaced OBs is a change from the original design and was suggested at the 2019 NA Kubecon storage face-to-face with the premise being that if there is no _technical_ reason for a cluster scoped resource then it should be namespaced.

As is true for PVCs-PVs, there is a 1:1 relationship between an OBC and an OB, and as will be seen below, there is a [_binding_](#binding) between an OBC and OB. 

OBCs reference a storage class. The storage class references the external provisioner, defines a reclaim policy, and specifies object-store specific parameters, such as region, owner secret, and bucket lifecycle polcies.
For brownfield provisioning, the storage class also specifies the name of the existing bucket so that app developers are not burdened with typically long, random bucket names.

> ### Moving Parts

Author: I will see if this can be done in the `Moving Parts` section

Before deploying an object bucket consuming app, the OB and OBC CRDs need to be created with RBAC rules granting _get_, _list_, and _watch_ verbs.
An object store needs to be created and referenced in the storage classes used by OBCs.

> Depending on the object store, a secret in the object-store's namespace, containing new bucket owner credentials, may need to be created.

Author: agree

The OBCs needed by the app should exist in the app's namespace.

> As is true for dynamic PV provisioning, a bucket provisioner pod needs to be running for each object-store created in the Kubernetes cluster.

Author: "operator" is somewhat of a red hat technique. Pod is accurate isn't it? It's true that this pod could do more than bucket provisioning -- is that your concern?

The secret contains the credentails needed to access the bucket and the configmap contains bucket endpoint information.
**Note:** it was mentioned at the 2019 NA Kubecon storage face-to-face that to improve scaling we could decide to move the configmap info into the secret, thus reducing k8s resources required.
The secret and configmap reside in the same namespace as the OBC.
The app pod consumes the secret and configmap and thus can reference the bucket.

An OBC for an existing bucket triggers the correct provisioner to grant access to the existing bucket. Like new bucket provisioning, associated bucket related artifacts (credentials, policy, etc) are typically created by the provisioner.

When a _greenfield_ OBC is deleted the associated provisioner is expected to delete the newly provisioned bucket and related object store-specific artifacts.
The library will delete the secret, configmap and OB.
A deleted _brownfield_ OBC is similar except that provisioners typically will not delete the physical bucket. Instead, bucket access is revoked and related artifacts are expected to be cleaned up.
Again, the library will delete the secret, configmap and OB.

> ### API

Reviewer 3: "API" is used interchangeably up to this point to reference both the k8s API (CRDs) and the library's interface.  How about defining terms for each at the start ("API" == CRDs, "Interface" == library methods)

There are four interface methods that must be defined by all bucket provisioners.
Additionally there are two required and one optional library functions that are used. 

#### Library Functions

- **`NewProvisioner`** is a required function called by provisioners to create the library's controller struct which is returned to the provisioner.
Each provisioner defines their own struct, passed to `NewProvision`, which implements the Interfaces below.
The returned struct supports the `Run` and `SetLabels` methods.

- **`Run`** is a required controller method called by provisioners to start the OBC controller.

> - **`SetLabels`** is an optional controller method called by provisioners to define the labels applied to the Kubernetes resources created by the library.

Reviewer 3: This could be shortened to "allows for setting custom labels on library objects"

  - "https://github.com/kube-object-storage/lib-bucket-provisioner/tree/master/doc/examples"
---

# Object Bucket Provisioning

## Table of Contents

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Design Details](#design-details)
  - [Custom Resources](#custom-resources)
  - [Moving Parts](#moving-parts)
  - [Interfaces](#interfaces)
  - [Library Usage](#library-usage)
  - [Current Restrictions](#current-restrictions)
  - [Risks and Mitigations](#risks-and-mitigations)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
    - [Examples](#examples)
      - [Alpha -&gt; Beta Graduation](#alpha---beta)
      - [Beta -&gt; GA Graduation](#beta---ga)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Implementation History](#implementation-history)
- [Drawbacks [optional]](#drawbacks-optional)
- [Alternatives [optional]](#alternatives-optional)
<!-- /toc -->

## Summary

Kubernetes natively supports dynamic provisioning for file and block storage but lacks support for object bucket provisioning.

> Without a formal API to describe and manage bucket provisioning, storage providers are left to write their own provisioners including all of the necessary controller code.

Reviewer 4: I'm uncomfortable with the implied idea that object bucket provisioning is just an extension of the existing block and file support.

Block and file support are accessed from within containers using local standard APIs.  Object storage is logically just a remote service - there's nothing in particular that makes it inherently similar to block/file.  Imagine that you wanted to be able to support provisioning against various "relational database as a service" products (e.g., RDS, CloudSQL, etc.).  You could imagine an api for things like "create a database" and "create credentials for accessing this database".  And you'd interact with the database via some sort of SQL client.  I see such a system as a much closer analog for object storage than block/file.

Author: Fair point. When I think of storage I think of file, block and object. I tend to lump object with file and block more than with databases because of its simplicity and narrower scope. 
Of course the fact that there is no mount for object makes it similar, from a pod's perspective, to a database. 
From a provisioner author's perspective I suppose there's no real difference between a bucket API and a DB API.

Author: I'll leave some time for others to chime in and then will push changes to this doc to reflect your comments.

Reviewer 3: > Of course the fact that there is no mount for object makes it similar, from a pod's perspective, to a database.

Agreed.  Given that the data path is so different between file/block and object, it's better to describe it as such.  We've already seen that by lumping object in with file/block, there's been some assumptions about whether we're also proposing a POSIX analog for bucket interfaces, which we're not.

This "wild west" landscape for bucket provisioning is potentially confusing, overly complex, inconsistent, may slow down Kubernetes adoption by object store users, and may hinder application portability.

We propose a common Kubernetes control-plane API for the management of object store bucket lifecycles which has no dependencies on the underlying providers.

## Motivation

By defining a Kubernetes API for object bucket provsioning, we provide a consistent framework for all object store providers and thus enhance the end user's storage experience. This API allows object store vendors to focus on how they wish to provision buckets instead of writing and testing the controller aspects of provisioning.

### Goals

+ Define a _control-plane_ object bucket management API thus relieving object store providers from Kubernetes controller details.

> + Minimize technical ramp-up for storage vendors who have already contributed external storage provisioners (file and block).

Reviewer 4: I don't understand why "who have already contributed..." is relevant.  Minimizing ramp-up is good, but again (per previous comment), I don't think there is any inherent association between object and volumes (block and file).

Author: Agree there is not necessarily an inherent association, per your comment above. Some storage vendor provide file, block and object store. My thinking was that there _could_ be some leverage if a file/block vendor decided to write their own bucket provisioner.

Reviewer 3: >  if a file/block vendor decided to write their own bucket provisioner.

The feedback we've gotten on the bucket provisioning library ("why not make it CSI-like?") makes even more sense when considering adoption by vendors into existing provisioners.  For one, CSI being the standard model for file/block means that it may actually be more cumbersome to incorporate the library.  Being written in go immediately creates more work for provisioner authors who don't use the language.  And the library pattern be in fact be unfamiliar to provisioner authors who jumped straight to CSI.

> + Make bucket consumption similar to file or block consumption using familiar concepts and commands.

Reviewer 4: I don't think it is bucket consumption that you are trying to make similar - I think it is bucket provisioning.

Reviewer 3: Right, it's the elements of the control plane that are meant to be familiar. It should not propose an abstraction of object store APIs specifically.

> + Use native Kubernetes resources where possible, again, to keep the bucket experience for users and admins similar to existing storage provisioning.

Reviewer 4: I'm assuming this refers to the decision to overload StorageClass for parameters.  I do not see the value in this - I see an anti-value.  Volume provisioning is fundamentally different than provisioning some external entity.  Artificial association is more likely to confuse IMO than reduce confusion.

Author: **This concern should be discussed**.
My team felt/feels it is important to use native storage classes for the reason cited. We know others who use a custom resource, similar to a storage class, but specific to their object store needs.

Reviewer 4: There are several benefits to a separate CR:

  * Avoiding cognitive dissonance - StorageClass is used for volume operations.  Other classes (e.g., VolumeSnapshotClass) have been introduced for other class-style parameterization.  Adding ObjectBucketClass would seem to be the most logical approach.  

  * You can evolve it independently of StorageClass, which responds to requirements from block/file (you don't have to bring them along, and you don't have to be brought along).

  * StorageClass is non-namespaced.  You've made progress towards making the object support install-able and usable without requiring write access to non-namespace objects.  If you had a single well-known namespace for object operations, ObjectBucketClass resources could go into that namespace (along with provisioners and OBs).

Reviewer 3: These are all pretty compelling points for a bespoke BucketClass object.  We leaned on SCs initially because of assumptions we made about the system being familiar to k8s users.  However, like you point out, we're beholden to the API which is exclusively intended for file/block.  Namespacing the classes with their provisioners also lends itself to scalability.

+ Be unopinionated about the underlying object-store and at the same time provide a flexible API such that provisioner specific features can be supported.
+ Ensure bucket consuming pods wait until the target buckets have been created and are accessible. Thus there is no specific order required for when a bucket claim is created vs. when the app pod is run.
+ Present similar user and admin experiences for both _greenfield_ (new) and _brownfield_ (existing) bucket provisioning.

### Non-Goals

+ Update the native Kubernetes PVC-PV API to support object buckets.
+ Define a native _data-plane_ object store API.
+ Handle the small percentage of apps that will not be portable due to use of non-compatible object-store features.

## Proposal

> Create a bucket provisioning library, similar to the Kubernetes [sig-storage-lib-external-provisioner](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go) library, which handles all of the controller and Kubernetes resources related to provisioning.

Reviewer 4: After years of dealing with the issues of having multiple vendors support provisioning of volumes, the community decided on an approach that abstracts the control plane behind a simple gRPC API, allowing for provisioning drivers to be effectively independent of Kubernetes.  I believe that this is the approach we should also use for object.  Instead of providing a library and asking object vendors to implement a controller using the library, I think we should provide top-level and sidecar controllers and define the interface via gRPC.  This massively simplifies the process of writing and deploying a driver.  Using a central top-level controller (e.g., see the 1.17 volume snapshot architecture) allows for a clean approach to CRD versioning and lifecycle management and standardizing observability metrics.

It also could theoretically allow for situations in the future where the gRPC functions could be optionally added to both CSI (for cases where volume storage providers *also* provide object support) and separately (e.g., via a new "OSI").

Author: **I'd like to see discussion on this suggestion.**
I am not experienced enough in CSI to offer arguments to this comment. There are a couple of provsioners using this lib and my understanding is that the effort was relatively low.

Reviewer 4: Having someone deliver the whole solution in a single controller creates a lot of problems.  For example:

  * If there is a bug in the library that is discovered, how do you ensure that all the controllers using that library are updated?

  * If you have multiple controllers installed, who "owns" the CRDs?  How do they get installed/upgraded?

  * If storage vendors are providing their own whole controllers, how do you ensure a degree of uniformity (and correctness) around observability?

The volume snapshot folks directly addressed this set of issues with the beta volume snapshot work and came up with a dual-control + driver architecture:

  * A single-instance global controller that handles most of the complicated creation, provisioner lookup, and binding logic

  * A very simple sidecar controller that serves primarily to convert k8s resources into gRPC calls

  * A plugin driver supplied by storage vendors that is kubernetes and language independent.

I'm advocating a very similar model here.

Reviewer 3: +1 And since many storage vendors have already switched over to CSI drivers, it's going to be a hard sell to convince them they should adopt a model that k8s has moved on from.

Author: I am starting to value a CSI-like approach (maybe COSI or "cozy?).
@alarge can you elaborate on this point:
> If there is a bug in the library that is discovered, how do you ensure that all the controllers using that library are updated?

I agree it's a potential problem with the library. But CRDs can support multiple versions and so can the lib (thought it doesn't today). Can you explain how CSI mitigates this issue?

Reviewer 14: @jeffvance A bug in this context could mean 2 kinds of bug: 1) kubernetes related bug(i.e CRD, controller common logic in the library) 2) object storage vendor related bug(i.e Provisioner bug). The library approach IMO seems to be unnecessarily coupling these 2 types of bugs. An issue on either side(K8s controller/CRD VS. Object Storage Provisioner) IMO shouldn't require attention from the other side. For example, a common library bug fix shouldn't require vendor to release a new version of their controller because there's no change in the "Provisioner" they've implemented. Having a CSI-like interface would allow us to decouple these 2 sides of the problem. For instance, a bug fix in the central controller would just require end-user to bump up the version number in the deployment yaml and object storage vendor is no involved at all.

Author: I do like the separation of provisioner bugs vs a csi bug. I'll add this point to the `CSI Pros` section of the KEP.

Author: Follow up: it [appears](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_client.go#L199) that `accessMode` is used by CSI. I don't yet know how it's enforced or controlled, but it is passed along in the request.

> This library will implement a simple and familiar API via Custom Resources, and define a _contract_ between app developers and provisioners regarding the Kubernetes resources related to bucket provisioning.

Reviewer 4: Per above, I think a gRPC API is a better form of contract than a golang-specific library interface.

Author: Is the main point here that gRPC has multi-language support?

Reviewer 4: Yes, that was part of my point.  But also the fact that the gRPC API is its own standalone thing, so documentation and version control tend to be handled better than internal library interfaces.

Since certain object store-specific artifacts, e.g. user name, are needed in order to clean up after a bucket is deleted or revoked, this proposal must allow for some provisioner values to be stored across API calls.

## Design Details

### Custom Resources

(see the [full specs](#custom-resource-details))

The bucket provisioning library utilizes two Custom Resources to abstract an object store bucket and a claim/request for such a bucket.
+ a namespaced _**ObjectBucketClaim**_ (OBC), similar to a PVC, is a formal request for a new or an existing bucket.
+ an _**ObjectBucket**_ (OB), similar to a PV, is the k8s representation of the provisioned bucket and may contain object store-specific data needed to allow the bucket to be de-provisioned. The OB resides in the same namespace as the object store, and is typically not seen by bucket consumers.
**Note:** namespaced OBs is a change from the original design and was suggested at the 2019 NA Kubecon storage face-to-face with the premise being that if there is no _technical_ reason for a cluster scoped resource then it should be namespaced.

> As is true for PVCs-PVs, there is a 1:1 relationship between an OBC and an OB, and as will be seen below, there is a [_binding_](#binding) between an OBC and OB. 

Reviewer 5: What is the reason to have two separate objects, OBC and OB? For PV/PVC it's security. If volume ID was in PVC, user could craft their own PVCs and guess volume ID of some other volume and steal its data. In this proposal, you give full bucket ID and secrets to access it to a pod. So user can guess ID of some other bucket and steal it in the pod.

Therefore you don't need OB object at all. All can be in OBC (and it could be called ObjectBucket instead of claim).

Author: A reason for the OB is to contain provisioner-specific data, eg. the newly created user name, so that provisioners can delete buckets and clean up policies etc.

Reviewer 5: Are these provisioner-specific data secret? In other words, can a malicious user misuse it to revoke / delete data of someone else?

Author: @jsafrane The secret is created by the lib and lives in the same namespace as the obc. The secret contains s3-like access and secret key values, and a provider-specific _map[string]string_.
The secret cannot be used maliciously AFAICT, because the provisioner is expected to return creds that are specific to a single bucket and are not the owner's creds.

Reviewer 3: @jsafrane This is dependent on the provisioner implementation.  The library only passes data returned from Provision()/Grant() calls directly into the secret.  Rook, for instance, creates a new Ceph user with limited rights and returns those credentials.  That user may only access the generated/attached bucket, may not create buckets, and (as a design of Ceph), may not view or mutate other user's buckets.

> OBCs reference a storage class. The storage class references the external provisioner, defines a reclaim policy, and specifies object-store specific parameters, such as region, owner secret, bucket lifecycle policies, etc.

Reviewer 4: Per previous comment, I think it would be better to have an ObjectBucketClass to handle these issues rather than overloading StorageClass with functionality not related to volumes.

### Moving Parts

Before deploying an object bucket consuming app, the OB and OBC CRDs need to be created with RBAC rules granting _get_, _list_, _watch_, and _delete_ verbs.
**Note:** OBCs do not need delete permission but OBs, secrets, and configmaps do.
An object store needs to be created and referenced in the storage classes used by OBCs.
The OBCs needed by the app must exist in the app's namespace.

As is true for dynamic PV provisioning, a bucket provisioner pod needs to be running for each type of object-store existing in the Kubernetes cluster.
For example, for AWS S3, the developer creates an OBC referencing a storage class which references the S3 store.

> This storage class needs to be created by the admin.

Reviewer 4: Note that using ObjectBucketClass instead of StorageClass may allow this class to be stored in a namespaced area along with the OBs

The S3 provisioner pod watches for OBCs whose storage classes point to the AWS S3 provisioner, while ignoring all other OBCs.
Likewise, the same cluster can also run the Rook-Ceph RGW provisioner, which also watches OBCs, only handling OBCs that reference storage classes which define ceph-rgw.

> **Note:** it is possible for one provisioner to handle OBCs for different instances of the same type of object store.

Reviewer 15: Wouldn't it be expected that a provisioner would handle OBCs for a given type of object store? Or this is just saying it's possible to have multiple provisioners for each type of object store?

Author: Example: a rook-ceph object store runs in namespace A and another rook-ceph object store runs in namespace B. A ceph-RGW provisioner runs in namespace C, where C could == A or B or any. Then that single provisioner could handle OBCs referencing storage classes which point to stores in A or B.

Reviewer 15: Ok, is this decision an implementation detail of the provisioner? In rook-ceph's case, I'm thinking for minimal privs the provisioner should only provision from the object store(s) in a given cluster. If you have multiple object stores in the same rook-ceph cluster (same namespace), you would have a single provisioner. But if you have multiple rook clusters (not as common), a separate provisioner may be used. If a single provisioner is needed, that provisioner needs RBAC access to all the rook-ceph clusters. Seems like it could be a deployment decision by the admin.

Author: It's really good to think about minimal privileges and, in fact, that's why this KEP proposes to make OBs namespaced rather than cluster scoped. However, provisioners need to be able to create secrets and configMaps in many namespaces (the ns of the OBC), and now need to create the OB in the ns of the store. If the admin wants a 1:1 mapping of object stores and provisioners s/he can limit write access to a single store. In that case, if an OBC references a storage class  which points to a store in a different ns (w/o RBAC rules) then the library's OB creation will fail and the lib's retry code will run.

This is a long way of saying that I should delete the note above since this is an admin detail.

Reviewer 3: > Wouldn't it be expected that a provisioner would handle OBCs for a given type of object store?

Yes.  It's just saying that a provisioner can handle different 'flavors' of OBCs that all reference the provisioner.  This would be represented by having more than 1 storage class for 1 object store where each SC encapsulates different feature sets.

#### Greenfield (new)

An OBC for a new bucket is defined by its inclusion of a generated or static bucket name.
In this case the referenced storage class omits a bucket name.
A new bucket OBC triggers the correct provisioner to create a new bucket and the associated artifacts (credentials, policy, etc). 
The library, in turn, creates a secret, configmap, and OB based on information returned by the provisioner.

The secret contains the credentails needed to access the bucket and the configmap contains bucket endpoint information.
**Note:** it was mentioned at the 2019 NA Kubecon storage face-to-face that to improve scaling we could decide to move the configmap info into the secret, thus reducing k8s resources required.
The OB is an abstraction of the bucket and can save some state data needed by provisioners.

> The secret and configmap reside in the same namespace as the OBC, and the OB lives in the provisioner's namespace.

Author: If a single provisioner for the same type of object store can handle this object store living in multiple namespaces, then the OB needs to reside in the object store's ns not the provisioner's ns.

> The app pod consumes the secret and configmap and thus can reference the bucket.

Reviewer 15: Multiple client pods can consume the same OBC to read/write to the same bucket, right?

Author: yes, subject to object store access rules. k8s has no opinion on this since there is no accessMode defined for buckets.

Reviewer 14: Is there any reason why we don't want accessMode input(opinion) from K8s side, presumably in claim? Read-only and Read-write access to buckets seem to be reasonable use cases. I'm imagining in a brownfield scenario, user can have a bucket setup for both consumer and producer application, accessMode would be handy to limit consumer to read-only access and allow producer to have read write access. Also read-only vs read-write seem to be well-supported across major object storage vendors.

Author: Agree that there are good use cases for `accessMode`. My (admittedly dated) understanding is that accessMode is unenforceable by k8s. Is this still true @jsafrane?
For bucket provisioning, the underlying object store would need to support various access modes.

Reviewer 5: Yes, AccessMode is used only during binding / provisioning, but it's never used when running a pod - you can have multiple pods using the same RWO volume in parallel.

> When a _greenfield_ OBC is deleted the associated provisioner is expected to delete the newly provisioned bucket and the related object store-specific artifacts.

Reviewer 15: There is a `retain` policy for the bucket, right? This makes it sound like the bucket is always deleted when the OBC is deleted.

Author: The retain policy is defined in the storage class and also stored on the OB. The `Delete` method passes the OB to the provisioner. The lib decides to call `Delete` rather than `Revoke` because the OBC is for a greenfield bucket. Provisioners have full control over whether or not to delete the bucket. The _expectation_ is that a  provisioner's `Delete` method will delete the bucket but it might not and still return a nil error.  Likewise, he _expectation_ is that a  provisioner's `Revoke` method will not delete the bucket but it might and still return a nil error.

Each provisioner will likely have to create and delete their own object store-specific resources.

Bucket binding requires these steps before the bucket is accessible to an app pod:
1. (greenfield only) generation of a random bucket name when requested (performed by bucket lib).
1. (greenfield only) the creation of the physical bucket with owner credentials (performed by provisioner).
1. creation of the necessary object store-specific resources, e.g. IAM user, policy, etc.
1. creation of an OB, residing in the object-store's namespace, based on the provisioner's returned OB structure (performed by bucket lib).
1. creation of a ConfigMap, residing in the OBC's namespace, based on the provisioner's returned endpoint info (performed by bucket lib).
1. creation of a Secret, residing in the OBC's namespace, based on the provisioner's returned credentials (performed by bucket lib).

### Custom Resource Details

#### OBC Details (created by user)

```yaml
apiVersion: objectbucket.io/v1alpha1
kind: ObjectBucketClaim
metadata:
  name: MY-BUCKET-1 [1]
  namespace: USER-NAMESPACE [2]
spec:

>   bucketName: [3]

Reviewer 4: Why should this be allowed? (claim specifying exact bucket name)

Reviewer 3: Do you mean as opposed to assigning a randomly generated name?  

To clarify, this field is ignored when specifying a brownfield storageClass (which I agree is a messy way of handling it, better suited to a BucketClass).

  generateBucketName: "photo-booth" [4]
  storageClassName: AN-OBJECT-STORE-STORAGE-CLASS [5]

>   additionalConfig: [6]

Reviewer 4: This feels like non-portable information, akin to passing through annotations.

Author: True. We want a way to pass provisioner specific information. The storage class (or bucketClass) allows this, but we were thinking of the case where, say region, was defaulted in the SC but could be overridden in the OBC. Of course each provisioner decides which of these fields to process or ignore.

1. additionalConfig gives providers a location to set store-specific config values (tenant, namespace...).
The value is a list of 1 or more key-value pairs.

**Note:** the OBC is mutated by the library with the addition of a finalizer and label:
```
  labels:
    bucket-provisioner: PROVISIONER-NAME [1]
  finalizers:
  - "objectbucket.io/finalizer" [2]
```
1. the label value is the name of the provisioner but due to Kubernetes restrictions slash (/) is replaced by a dash (-).
1. finalizers set and cleared by the lib's OBC controller. Prevents accidental deletion of an OB.

#### OB Details (created by library)

```yaml
apiVersion: objectbucket.io/v1alpha1
kind: ObjectBucket
Metadata:
  name: obc-my-ns-my-bucket [1]
  namespace: namespace-of-object-store [2]
  labels:
    bucket-provisioner: PROVISIONER-NAME [3]
  finalizers:
  - "objectbucket.io/finalizer" [4]
spec:
  storageClassName: example-obj-prov [5]
  claimRef: *v1.objectreference [6]
  reclaimPolicy: {"Delete", "Retain"} [7]
  endpoint:
    bucketHost: foo.bar.com
    bucketPort: 8080
    bucketName: my-photos-1xj4a
    region: # provisioner dependent
    subRegion: # provisioner dependent

>     additionalConfigData: [] #string:string

Reviewer 4: This list of endpoint parameters feels specific to a particular provider.  I feel like the goal of this should be to produce the fields that are needed by client SDKs and to be useful across at least the three main object APIs (S3, GCS, and Azure).   Given that, I'd argue that a "data path protocol" should be supplied with the claim (and declared by drivers).  The endpoint and credential information supplied by the driver (and recorded in the OB and Secret) should be appropriate for the given data path protocol (e.g., nearly directly usable by the standard API for that data protocol).

Reviewer 3: +1 It sounds like you're describing a pattern similar to the *VolumeSource subfield of PVs, which I agree would be a better fit.  The overuse of string:string maps is something we need to address generally.

Author: @alarge:
> "data path protocol" should be supplied with the claim (and declared by drivers).

A goal was to not embed specific object store fields into the lib (though admittedly there is an s3 flavor). The reason for this is so object store vendor would not have to wait for their fields to be added. IOW, the lib would have some `switch dataPathProtocol` block which has to know which fields go with which protocol -- at least that's how I'm seeing. Can you elaborate more on this?

1. namespce of the originating OBC.
1. finalizers set and cleared by the lib's OBC controller. Prevents accidental deletion of the Secret.
1. the library adds a label (seen here) but each provisioner can supply their own labels.
1. the label value shown is the name of the provisioner but due to Kubernetes restrictions slash (/) is replaced by a dash (-).
1. ownerReference makes this secret a child of the originating OBC for clean up purposes.
1. ACCESS_KEY_ID and SECRET_ACCESS_KEY are the only secret keys defined by the library.
Provisioners are able to cause the lib to create additional keys by returning  the `AdditionalSecretConfig` field.
**Note:** the library will create the Secret using `stringData:` and let the Secret API base64 encode the values.

#### ConfigMap (created by library)

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: MY-BUCKET-1 [1]
  namespace: OBC-NAMESPACE [2]
  finalizers: [3]
  - objectbucket.io/finalizer
  labels: [4]
    bucket-provisioner: aws-s3.io-bucket [5]
  ownerReferences: [6]
  - name: MY-BUCKET-1
    ...
data: 
  BUCKET_HOST: http://MY-STORE-URL [7]
  BUCKET_PORT: 80 [8]
  BUCKET_NAME: MY-BUCKET-1 [9]
  BUCKET_REGION: us-west-1

>   ... [10]

Reviewer 4: Pre previous comment - these should be whatever is appropriate for the given data path protocol.

Author: If we do this and have a "DataPathProtocol" type field in the OBC then app pods have little chance of being portable across different object store, right? Some degree (obviously not 10%) of portability is/was desired in the original design.

Reviewer 3: The app would be tethered to the protocol regardless as it's expected they've been written using some SDK.   This doesn't restrict apps to a specific object provider, but they do need to be aware of what protocol is being used at the endpoint.  That is, S3 is pretty common, but not ubiquitous, and this must be communicated.

1. **all** parameter keys and values are specific to a provisioner, are optional, and are not validated by the StorageClass API.
Fields to consider are object-store endpoint, version, possibly a secretRef containing info about credential for new bucket owners, etc.
1. bucketName is required for access to existing buckets.
Unlike greenfield provisioning, the brownfield bucket name appears in the storage class, not the OBC.
1. each provisioner decides how to treat the _reclaimPolicy_ when an OBC is deleted. Supported values are:
+ _Delete_ = (typically) physically delete the bucket.
Depending on new vs. existing bucket, the provisioner's `Delete` or `Revoke` methods are called.
+ _Retain_ = (typically) do not physically delete the bucket.
Depending on the provisioner, various clean up steps can be performed, such as deleting users, revoking credentials, etc.
For both new and existing buckets the provisioner's `Revoke` method is called.

#### App Pod (independent of provisioner)

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: app-pod
  namespace: dev-user
spec:
  containers:
  - name: mycontainer
    image: redis

>     envFrom: [1]

Reviewer 4: This is just an example, right?  Mounted secrets might be preferred (e.g., in the case of multiple bucket claims)

Reviewer 3: Yes.  It would probably be helpful to add that case as well.

    - configMapRef: 
        name: MY-BUCKET-1 [2]
    - secretRef:
        name: MY-BUCKET-1 [3]
```
1. use `env:` if mapping of the defined key names to the env var names used by the app is needed.
1. makes available to the pod as env variables: BUCKET_HOST, BUCKET_PORT, BUCKET_NAME

> 1. makes available to the pod as env variables: ACCESS_KEY_ID, SECRET_ACCESS_KEY

Reviewer 4: Again, specific to S3.  Different variables for different data path protocols.

+ Use standard Kubernetes mechanisms to sync a pod with the readiness of the bucket it will consume. This can be accomplished via Secrets.

## Non-Goals
+ Define a native _data-plane_ object store API which would greatly improve object store app portability.
+ Mirror the static workflow of PersistentVolumes wherein users are given the first available Volume.  Pre-provisioned buckets are expected to be non-empty and thus unique.

##  Vocabulary

+  _Brownfield Bucket_ - externally created and represented by a `BucketClass` and managed by a provisioner.
+ _Bucket_ - A user-namespaced custom resource representing an object store bucket.
+  _BucketClass_ - A cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set.
   + _In Greenfield_: an abstraction of new bucket provisioning.
   + _In Brownfield_: an abstration of an existing objet store bucket.
+ _BucketContent_ - A cluster-scoped custom resource bound to a `Bucket` and containing relevant metadata.
+ _Container Object Storage Interface (COSI)_ -  A specification of gRPC data and methods making up the communication protocol between the driver and the sidecar.

> + _COSI Controller_ - A central controller responsible for managing `Buckets`, `BucketContents`, and Secrets.

Author: If we stick with the original implementation details of this KEP (now outdated due to narrowing the focus strictly to the API), then there is a single COSI controller, and a sidecar controller shared by all provisioners. Deployments of these controllers could be done by an operator or manually, I'd think.

> + _Driver_ - A containerized gRPC server which implements a storage vendor’s business logic through the COSI interface. It can be written in any language supported by gRPC and is independent of Kubernetes.

Reviewer 16: Why is a grpc interface needed?  Just use a controller.

Grpc plug-ins lack many desirable characteristics of controllers and can be almost completely eliminated for the sorts of operations described here.

Author: Our motivation for using gRPC is similar to CSI's reasoning: 1) allow storage vendor provisioning to be container orchestrator agnostic. 2) allow storage vendors to write their provisioners in a variety of languages with the potential that the same driver could run in different orchestrators. 3) separate patching/versioning of vendor-specific code from k8s controller/sidecar code (which is difficult if a library is to be imported by the vendor).
The downsides are 1) there are separate containers now: a) the driver itself, b) a sidecar container mapping k8s/go structs to gRPC services/messages, c) a central controller. 2) it's more complex than a single controller.

What controller characteristics are we foregoing with this proposal? Thanks.

Reviewer 16: I don't think CSI needed to use GRPC except at a node level, and honestly most of the node level things are using GRPC because we don't have good support for controller apis at the node level.  That's not a reason to use GRPC otherplaces.

Yeah, I would say 1 doesn't seem to be the right approach (generic things to container orchestrators is less critical to success of kube than it used to be).

2 you can write controllers in lots of languages, but supporting lots of languages is not really a technical design goal of good software in general, it is more specific than that (sometimes supporting lots of languages is valuable).  How many CSI drivers are in a non-go language?  How many CRI runtimes?  Either way, controllers can be in many languages and that is just as important a goal as anything else.

3. I don't buy the argument that vendoring kube code is hard - it was 4 years ago, it shouldn't be now.  Our ecosystem is much expanded.

This is the classic controller pattern - talk to a backend, provision something, update an api object.  You can collaborate with multiple controllers just as easily as with GRPC plugins.  You reduce the total stack complexity.  You can define a stock Go interface that pepole can fork.  You can describe the pattern for a third party.

I am very anti-GRPC style plugins when controllers solve the problem equally well.  We have allowed it in auth plugins and a few node level things, but it points to gaps in tooling we have mostly closed at this point.  Most plugins need to make kube api calls, so they aleady inherit most of a controller.

Reviewer 16: Just reviewing the KEP, i see nothing that requires a gRPC plugin, I just see a bog standard kubernetes controller (which is our preferred mechanism for 3rd party extension).  Having an optional GRPC plugin for a small set of vendors sounds great, but since every major cloud and storage platform I'm aware of already has a go object storage bucket API (either as part of an operator or using a small integration), using go does not seem to be a blocker.

The dominant pattern should be a set of base code and Go interfaces for implementing those with the ability to both keep the core libraries up to date and fork those to provide additional behavior.

Reviewer 3: Thanks @smarterclayton.  These are great points.  1 and 2 I agree with outright.

> I don't buy the argument that vendoring kube code is hard - it was 4 years ago, it shouldn't be now. Our ecosystem is much expanded.

IMO, this has been the biggest reason we've been amicable towards gRPC over a go interface.  We went down this path in a [previous incarnation](https://github.com/kube-object-storage/lib-bucket-provisioner) of a bucket provisioning API.  

The issue isn't that importing k8s tools is hard, it's when multiple dependencies wraps k8s tools, they inevitably result in a dep hell.  Go mods alleviate this for the most part but I don't think we can expect existing projects, particularly those written prior to mods, to _not_ encounter the same pain we did before.

> The dominant pattern should be a set of base code and Go interfaces for implementing those with the ability to both keep the core libraries up to date and fork those to provide additional behavior.

What I think you're suggesting is a that _instead_ of a library, a standalone controller be provided along with a go interface.  In effect, merging the sidecar and provisioner code under a single binary.  The controller manages the APIs and executes provisioner methods, deployed alone in it's one container.  Am I close to the mark?

  
## System Configuration

+ The COSI controller runs in the `cosi-system` namespace where it manages `Buckets`, `BucketContents`, and Secrets. This namespace name is not enforced but suggested.
+ The Driver and Sidecar containers run together in a Pod and are deployed in any namespace, communicating via the Pod's internal network (_localhost:some-port_). We expect and will document that different drivers live in separate namespaces.
+ Operations must be idempotent in order to handle failure recovery.

### Unique Driver Names

It is important that driver names are unique, otherwise multiple sidecars would try to handle the same `BucketContent` events (since the sidecar matches on driver name).   The prescribed pattern  to be used for all provisioner names

## Workflows

### Determining Case from BucketClass

| BucketClassFields             | SecretRef: non-nil | SecretRef: nil |
| ----------------------------- | ------------------ | -------------- |
| **BucketContentRef: non-nil** | Static             | Brownfield     |
| **BucketContentRef: nil**     | Undefined          | Greenfield     |

> #### Create Bucket (Greenfield)

Reviewer 15: `BucketContent` is confusing to me.  If you talk about the contents of a bucket, you're referring to the actual data objects held by the bucket. Similarly, you don't talk about a PV's content, you just talk about the PV. We should just talk about the "Bucket". What about these names?
- `BucketRequest` (or `BucketClaim`): The app's request for a bucket (renamed from `Bucket`)
- `Bucket`: Used for internally tracking the bucket creation/deletion (renamed from `BucketContent`)
- `BucketClass`: unchanged

Author: I (personally) like `BucketRequest` because that's how we often refer to it in conversations since "bucket" is ambiguous. I also like OBC but "claim" is not in favor with several folks. I prefer `BucketObject` or `OB` over `Bucket` due to the same ambiguous argument. When we say "bucket" it's sometimes hard to know if we mean the k8s resource or the physical bucket.

1. Sidecar creates a `secret` containing the endpoint and credentials, with a random/unique name and `ownerRef` set to `BucketContent`.
1. Sidecar updates `BucketContent.secretRef` with its `secret` name and namespace and sets `BucketContent.status.phase` to *“Ready”*.
1. Controller detects the `BucketContent` update and sees the *“Ready”* phase. 
1. Controller copies the generated `secret` to the `Bucket` namespace with name `Bucket.secretName`.  The `secret` is created with an ownerRef of the Bucket.
1. Controller *“binds”* the `Bucket` and `BucketContent` by setting `bucket.status.bucketContentName` and `bucketContent.bucketRef`, and sets both statuses to *“Bound”*.
1. App `pod` ingests `secret` and runs.

#### Delete Or Revoke Access (Greenfield & Brownfield)

1. The user deletes their `Bucket`, which blocks until the `finalizer` is removed.
1. The Controller detects the event and sees the `deletionTimestamp` set in `Bucket` and gets the `BucketClass`. 
1. If the `BucketClass.secretRef` is nil, the object store bucket is not static, and the process continues to step 4.

> 1. The Controller deletes the referenced `BucketContent` object, which blocks until the `finalizer` is removed.

Reviewer 15: Deleting a bucket can be a long-running operation if there are many objects in the bucket. We should mark the underlying bucket for deletion and allow it to be removed asynchronously IMO. Is this the proposal, or would it be up to each driver to implement whether it blocks until the bucket and all its contents are removed?

Author: I think it would be up to each driver (assuming we implement earlier versions of this KEP where we discussed automation implementation ideas). But this is an interesting point and probably needs further discussion after MVP. @travisn do you think API changes will be needed if we want to support async delete later on?

1. The Sidecar detects the `BucketContent` event and sees the `deletionTimestamp`, and gets the referenced `BucketClass`.
1. If the `BucketClass.bucketIdentifier` is nil, the Sidecar decides the `BucketClass.bucketIdentifier` is a greenfield object store bucket and calls the rpc associated with the `BucketClass.releasePolicy` (*DeleteBucket* or *RevokeBucketAccess*). Otherwise, the `BucketClass.bucketIdentifier` is non-nil, indicating that it is a brownfield object store bucket, and the Sidecar calls the *RevokeBucketAccess()* rpc.
1. The Sidecar sets `BucketContent.status.phase` to *“Released”.*
1. The Controller sees `BucketContent` status is “*Released*” and removes `BucketContent`’s `finalizer`.
1. The `BucketContent` and the dependent `Secret` will be garbage collected.
1. The Controller removes the `finalizer` from the `Bucket`, allowing it and it’s dependent `secret` to be garbage collected.

### Static Buckets

> > Note: No driver, and thus no sidecar, are present to manage provisioning.

Reviewer 15: Are static buckets not just a special case of brown field? Or how is it different?

Author: Yes. The difference is there is no driver/sidecar so there are more manual steps. The idea with static/driverless is to encourage early COSI adoption where there may not be drivers avail.

- As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to object storage instances.

## APIs

### Storage APIs

#### BucketRequest

A user facing, namespaced custom resource requesting a bucket endpoint. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is required in order to grant credentialed access to the bucket. 

**NOTE:** _Currently under review: we are discussing whether or not a BR is needed for any brownfield use-cases. We are considering BRs only apply to greenfield bucket creation (where a BAR is also needed), and that all brownfield accesses need only a BAR._

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketRequest
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
  - cosi.io/finalizer [2]
spec:
  protocol: [3]
  bucketPrefix: [4]
  bucketClassName: [5]
  bucketInstanceName: [6]
status:

>   phase: [7]

Reviewer 9: According to API conventions (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties), `The pattern of using phase is deprecated. Newer API types should use conditions instead. `

Author: Thanks @xing-yang !  From this link:
```
The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted system-design principles and hampered evolution, since adding new enum values breaks backward compatibility. Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See #7856 for more details and discussion.
```

> Note: the `Bucket`'s release policy is set to "Retain" as a default. Exercise caution when using the "Delete" release policy as the bucket content will be deleted.
> Note: a `Bucket` is not deleted if it is bound to any `BucketRequest`s.
6. `anonymousAccessMode`: a string specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the storage instance or objects are intended to be publicly readable and/or writable. One of:
   - "private": Default, disallow uncredentialed access to the storage instance.
   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
> Note: does not reflect or alter the backing storage instances' ACLs or IAM policies.
7. `bucketClassName`: Name of the associated bucket class (greenfield only).
8. `bindings`: an array of `BucketAccess.name`(s). If the list is empty then there are no bindings (accessors) to this `Bucket` instance and the `Bucket` can potentially be deleted.

> 9. `protocol`: The protocol the application will use to access the storage instance.

Reviewer 6: I believe this design makes it impossible for a single `Bucket` (its objects) to be exposed using multiple protocols. Is this intentional? What's the suggested approach for storage systems which support multi-protocol access to the same buckets?

Reviewer 8: There is no object storage provider that provides multiple protocol support for any single bucket. The only way to access a single bucket with different protocols is by using a protocol-translating proxy. eg. S3Proxy

In case the proxy is being utilized, multiple bucket objects can be created for each of the protocols, and have them point to their respective proxies.

Reviewer 6: > There is no object storage provider that provides multiple protocol support for any single bucket. The only way to access a single bucket with different protocols is by using a protocol-translating proxy. eg. S3Proxy

This is incorrect. Our solution allows a single bucket and its object to be accessed using either an AWS S3-compatible API, or the Azure Blob-compatible API, with authn/z interop, without `s3proxy` or something similar. As such, this design limitation directly affects/limits creating large-scale applications where some services use one API, others another, to access the same datastore.

   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
     - `s3`: data required to target a provisioned S3 bucket and/or user.
     - `gcs`: data required to target a provisioned GCS bucket and/or service account.

> 10. `allowedNamespaces`: a copy of the `BucketClass`'s allowed namespaces. Additionally, this list can be mutated by the admin to allow or deny namespaces over the life of the bucket.

Reviewer 6: It is, in general, not very common, and definitely 'clumsy' to work with object UIDs. Would it make sense to explicitly state (in this document) why this field is not a set of `Namespace` names, but those UIDs are required?

Reviewer 8: This is intended to be a access restriction mechanism. If a namespace from the `allowedNamespaces` list is deleted and another namespace added by the same name, then the new namespace should not have access to this Bucket.

Reviewer 6: That's what I thought. Will require good UX though to be usable, since AFAIK there's no other place in K8s where one is required to provide UIDs directly.

A cluster-scoped custom resource to provide admins control over the handling of bucket provisioning.  The `BucketClass` (BC) defines a release policy, specifies driver specific parameters, and provides the provisioner name. A list of allowed namespaces can be specified to restrict new bucket creation and access to existing buckets. A default bucket class can be defined for each supported protocol, which allows the bucket class to be omitted from a `BucketRequest`. Most of the `BucketClass` fields are copied to the generated `Bucket` instance.

> Note: a `BucketClass` is immutable, like a storage class, but the fields copied to the `Bucket` can be edited by the admin.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketClass
metadata:
  name: 
provisioner: [1]
isDefaultBucketClass: [2]
protocol: {"azureblob", "gcs", "s3", ... } [3]
anonymousAccessMode: [4]
releasePolicy: {"Delete", "Retain"} [5]
allowedNamespaces: [6]
  - name:
    uid:
parameters: [7]
```

1. `provisioner`: the name of the driver. If supplied the driver container and sidecar container are expected to be deployed. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".

> 2. `isDefaultBucketClass`: (optional) boolean, default is false. If set to true then potentially a `BucketRequest` does not need to specify a `BucketClass`. If the greenfield `BucketRequest` omits the `BucketClass` and a default `BucketClass`'s protocol matches the `BucketRequest`'s protocol then the default bucket class is used.

Reviewer 6: What happens when there are two default `BucketClass`es for the same `protocol`? If this is never supposed to happen, who (which admission controller, I guess) is responsible for this to be the case?

Reviewer 8: The COSI central controller will register webhooks for this. This is not currently shown since we want to minimize the number of changes before the next KEP review. This adminssion web hook does not affect the overall architecture.

Reviewer 6: My point is mainly: the spec ought to define how the system behaves in various states of the system, or which invariants are to be kept (e.g., only a single default BC per protocol). Indeed, how this is enforced in the implementation is a design decision and should not necessarily be defined in this spec.

1. `bucketRequestName`: the name of the `BucketRequest` associated with this access request. From the bucket request, COSI knows the `Bucket` instance and thus bucket and its properties.
1. `bucketAccessClassName`: name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.
1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance.
1. `conditions`: the latest available observation of this `BucketAccessRequest`. (Phase is not used.)
1. `Type`: only one type is supported: "AccessGranted". If the condition's `Status` is true then access to the bucket has been granted for the `Reason` provided. If `Status` is false then the bucket is unable to be accessed and, again, `Reason` states why.
1. `Status`: one of true, false or unknown. If not specified then unknown is assumed. A true status indicates the this condition is true at the time of observation.
1. `Reason`: one of "Bound", "Granting", "Revoking".
1. `Message`: optional human readable message containing more details.

#### BucketAccess

> A cluster-scoped administrative custom resource which encapsulates fields from the `BucketAccessRequest` (BAR) and the `BucketAccessClass` (BAC).  The purpose of the `BucketAccess` (BA) is to serve as communication path between provisioners and the central COSI controller.  In greenfield, the COSI controller creates `BucketAccess` instances for new `BucketAccessRequest`s. In brownfield the admin is expected to manually create the BA. There is one `BucketAccess` instance per `BucketAccessRequest`.

Reviewer 6: > There is one `BucketAccess` instance per `BucketAccessRequest`.

What wasn't immediately clear to me from this sentence, but *is* from the schema of the object, is that a single BA is also tied to a single BAR. The sentence above may be interpreted as "A BAR has one BA, but a BA can be tied to multiple BARs", which is not true.

Given that, I wonder why BAs must be cluster-scoped?

Reviewer 8: We did not want credentials to live in the application namespace.

1. `policyActions`: a set of provisioner/platform defined policy actions to allow or deny a given user identity.
1. `parameters`: (Optional)  A map of string:string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.

---

### App Pod
The application pod utilizes CSI's inline empheral volume support to provide the endpoint and secret credentials in an in-memory volume. This approach also, importantly, prevents the pod from launching before the bucket has been provisioned since the kubelet waits to start the pod until it has received the cosi-node-adpater's `NodePublishVolume` response.

Here is a sample pod manifest:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-app-pod
  namespace: dev-ns
spec:
  serviceAccountName: [1]
  containers:
    - name: my-app
      image: ...
      volumeMounts:
        - mountPath: /cosi [2]
          name: cosi-vol
  volumes:
    - name: cosi-vol
      csi: [3]
        driver: cosi.sigs.k8s.io [4]
        volumeAttributes: [5]
          bucketRequestName: <my-br-name>
          bucketAccessRequestName: <my-bar-name>

>           bucketRequestNamespace:

Reviewer 6: Does this mean a Pod can request/get access to a BR/BAR from another Namespace? That's a bit surprising.

Reviewer 8: Good catch, this should not be here. @jeffvance --^

Author: II believe the cosi adapter needs this namespace. @wlan0 ? The key name could be more generic if it's confusing to see a BAR ns but not the BR ns (they are the same).

Reviewer 6: Can't you rely on CSI's `podInfoForMount` feature (https://kubernetes-csi.github.io/docs/pod-info.html) to retrieve this value?

+ COSI gets the `BR.BucketClass` (directly or via the matching default).
+ COSI gets the `BAR.bucketAccessClass` object.
+ COSI creates the associated `Bucket`, after filling in a template consisting of fields from the `BucketClass` and BR.
  + If this create operation fails due to the `Bucket` already existing, then we have a brownfield case; otherwise it's greenfield.  **Note:** the goal here is to reduce separate greenfield vs. brownfield logic in the code.
+ COSI creates the associated `BucketAccess` (BA), after filling in a template consisting of fields from the `BucketAccessClass` and BAR.
+ For a newly created `Bucket`, the sidecar sees it and gRPC calls the driver to create a bucket.
+ COSI is done with the `BucketRequest`, but there is no binding until the `BucketAccess` instance is created for the associated `BucketAccessRequest`.
+ The sidecar sees the newly created BA and gRPC calls the driver to grant access.
+ COSI sees the completed BA which triggers binding the BA in `Bucket`.

+ Depending on when the app pod was started, the kubelet call `NodePublishVolume` and waits for the response from the cosi-node-adapter.

> + The cois-node-adaper sees the `NodePublishVolume` request and is passed the `BucketRequest` and `BucketAccessRequest` names and namespace.

Reviewer 6: ```suggestion
+ The cosi-node-adaper sees the `NodePublishVolume` request and is passed the `BucketRequest` and `BucketAccessRequest` names and namespace.
```

+ The adapter gets the corresponding `Bucket` and `BucketAccess` instances, and verifies that the BA has been bound.
+ The adapter creates the host files for the secret and endpoint info.

> + At this point the adapter responds to the `NodePublishVolue` request and the kubelet continues to launch the pod

Reviewer 6: ```suggestion
+ At this point the adapter responds to the `NodePublishVolume` request and the kubelet continues to launch the pod
```

##### Sharing Dynamically Created Buckets (green-brown)
Once a `Bucket` is created, it is discoverable by other users in the cluster (who have been granted the ability to list `Bucket`s or via non-automated methods).  In order to access the `Bucket`, a user must create a `BucketRequest` (BR) that specifies the `Bucket` instance by name. This `BucketRequest` should not specify a `BucketClass` since the `Bucket` instance already exists.
The user also needs to creates a `BucketAccessRequest` (BAR), which references the BR. At this point the workflow is the same as above.

### DeleteBucket

![DeleteBucket Workflow](COSI%20Architecture_Delete%20Bucket%20Workflow.png)

> _Delete_ covers deleting a bucket and/or revoking access to a bucket. A `Bucket` delete is triggerd by the user deleting their `BucketRequest`. A `BucketAccess` removal is triggered by the user deleting their `BucketAccessRequest`. A bucket is not deleted if there are any bindings (accessors). Once all bindings have been removed the `Bucket`'s `conditions` is marked unavailable, **and** if the release policy is "Delete", then the sidecar will gRPC call the provisioner's _Delete_ interface. It's up to each provisioner whether or not to physically delete bucket content, but the expectation is that the physical bukcet will at least be made unavailable.

Reviewer 6: ```suggestion
_Delete_ covers deleting a bucket and/or revoking access to a bucket. A `Bucket` delete is triggered by the user deleting their `BucketRequest`. A `BucketAccess` removal is triggered by the user deleting their `BucketAccessRequest`. A bucket is not deleted if there are any bindings (accessors). Once all bindings have been removed the `Bucket`'s `conditions` is marked unavailable, **and** if the release policy is "Delete", then the sidecar will gRPC call the provisioner's _Delete_ interface. It's up to each provisioner whether or not to physically delete bucket content, but the expectation is that the physical bucket will at least be made unavailable.
```

    rpc ProvisionerDeleteBucket (ProvisionerDeleteBucketRequest) returns (ProvisionerDeleteBucketResponse) {}

    rpc ProvisionerGrantBucketAccess (ProvisionerGrantBucketAccessRequest) returns (ProvisionerGrantBucketAccessResponse);
    rpc ProvisionerRevokeBucketAccess (ProvisionerRevokeBucketAccessRequest) returns (ProvisionerRevokeBucketAccessResponse);
}
```

#### ProvisionerGetInfo

This call is meant to retrieve the unique provisioner Identity. This identity will have to be set in `BucketRequest.Provisioner` field in order to invoke this specific provisioner.

```
message ProvisionerGetInfoRequest {
    // Intentionally left blank
}

message ProvisionerGetInfoResponse {    
    string provisioner_identity = 1;
}
```

#### ProvisonerCreateBucket

This call is made to create the bucket in the backend. If the bucket already exists, then the appropriate error code `ALREADY_EXISTS` must be returned by the provisioner.

```
message ProvisionerCreateBucketRequest {    
    // This field is REQUIRED
    string bucket_name = 1;

    map<string,string> bucket_context = 4;

    enum AnonymousBucketAccessMode {
	BUCKET_PRIVATE = 0;

> 	BUCKET_READ_ONLY = 1;

Reviewer 6: These now no longer match the values defined at https://github.com/kubernetes/enhancements/pull/1383/files#diff-9942e2f090a0582218253969881d6409R210. Could make sense to keep them aligned, for clarity.

+ _driverless_ - a system where no driver is deployed to automate object store operations.  COSI automation may still be deployed to managed COSI APIs. **Note:** the current state of the KEP does not justify nor define driverless aka "static provisioning". If the community feels this is an MVP requirement we will need to flush this out.

# Proposal

## User Stories

#### Admin

- As a cluster administrator, I can control access to new and existing buckets when accessed from the cluster, regardless of the backing object store.

#### User

- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
- As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to object storage instances.

## APIs

### Storage APIs

#### BucketRequest

> A user facing, namespaced custom resource requesting a bucket endpoint. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentialed access to the bucket. BRs are required for bo th greenfield and brownfield uses.

Reviewer 6: ```suggestion
A user facing, namespaced custom resource requesting a bucket endpoint. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentialed access to the bucket. BRs are required for both greenfield and brownfield uses.
```

For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.

```yaml
apiVersion: cosi.io/v1alpha1
kind: Bucket
metadata:
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
    - cosi.io/finalizer [3]
spec:
  provisioner: [4]
  releasePolicy: [5]
  anonymousAccessMode: [6]
    - private
    - publicRead
    - publicReadWrite
  bucketClassName: [7]
  bindings: [8]
    - "<BucketAccess.name>"
  protocol: [9]
    protocolSignature: ""
    azureBlob:
      containerName:
      storageAccount:
    s3:
      endpoint:
      bucketName:
      region:
      signatureVersion:
    gcs:
      bucketName:
      privateKeyName:
      projectId:
      serviceAccount:
  allowedNamespaces: [10]
    - name:
      uid:
  parameters: [11]
status:
  conditions: [12]
    - Type: {BucketAvailable} [13]
      Status: [14]
      Reason: [15]
      Message: [16]
```

> 1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"bucket-"<bucketRequest.name>"-"<bucketRequest.namespace>_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name.

Reviewer 6: Any particular reason for the `bucket-` prefix? Names in K8s are limited in length, and I believe it's pretty sure this object is... a bucket.

Author: made same change to `BucketAccess.name`.

> 1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"bucket-"<bucketRequest.name>"-"<bucketRequest.namespace>_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name.

Reviewer 6: Wouldn't `namespace-name` iso `name-namespace` make more sense, e.g. when ordering a `BucketList`, when (tab-)completing in some CLI/UI,...

3. `finalizers`: added by the controller to defer `Bucket` deletion until backend deletion ops succeed.
4. `provisioner`: The provisioner field as defined in the `BucketClass`.  Used by sidecars to filter `Bucket`s. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
5. `releasePolicy`: Prescribes outcome of a Delete events. 
   - _Retain_: (default) the `Bucket` and its data are preserved. The `Bucket` can potentially be reused.
   - _Delete_: the bucket and its contents are destroyed.
> Note: the `Bucket`'s release policy is set to "Retain" as a default. Exercise caution when using the "Delete" release policy as the bucket content will be deleted.
> Note: a `Bucket` is not deleted if it is bound to any `BucketRequest`s.
6. `anonymousAccessMode`:  ACL specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the storage instance or objects are intended to be publicly readable and/or writable.  Accepted values:
   - `private`: Default, disallow uncredentialed access to the storage instance.

>    - `ro`: Read only, uncredentialed users can call ListBucket and GetObject.

Reviewer 6: This seems not in line with the YAML above, which lists e.g., `publicReadWrite`, not `rw`.

Author: Good catch! I've changed this to a simple string with a single value and made it consistent with bucket class.

+ _static provisioning_ - custom resource creation is done by the admin rather than via COSI automation. This may also include _driverless_ but they are independent.

# Proposal

## APIs

### Storage APIs

#### BucketRequest

A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.

There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.

```
apiVersion: cosi.io/v1alpha1
kind: BucketRequest
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
    - cosi.io/finalizer [2]
spec:
  protocol: [3]
  bucketPrefix: [4]
  bucketClassName: [5]
  bucketInstanceName: [6]
status:
  bucketAvailable: [7]
```

> 1. `labels`: added by COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

Reviewer 11: Can a user provide labels to be read by the provisioner?

Author: Yes. The user can add labels to their BRs and BARs. The admin can add labels to the BCs and BACs, But we do not have a plan to allow the admin to tell COSI to add labels to the generated B and BA. That would be nice but probably after MVP.

Reviewer 11: OK, so I was trying to make sure that the provider code gets the BR/BAR labels when processing the grpc request, so that it will be possible for users to ask for custom configurations from the provisioner (which will take it's own decision to allow or reject). Can we make sure those labels are sent over grpc?

Author: That's an interesting point. Basically should the user, through their BR, be able to influence characteristics of the new bucket -- in essence overriding or adding to the BC? 
IMO, if we allow this it should be explicit and not done via labels.
Also, at one point on the KEP we allowed a BR to specific a map for bucket properties. The BC also had this map (separate from _parameters_). COSI would have merged the BR's map with the BC's map such that the BC's overwrote the same key in the BR. This idea was rejected.

Reviewer 11: I know it was rejected, this is why I don't suggest to standardize this feature, but I just want to make sure that COSI will enable providers to bypass the limitation until standards will follow.

Author: Whether the mechanism is standard or formal we want to maintain a separation of concern between the user and the admin. Passing labels k-v to the driver can potentially allow a user to override a `BucketClass.parameters` values, or add new values that were omitted in the BC. My vote is to _not_ merge labels into the `Bucket.parameters` field (which is passed as bucket_context to the driver).

Reviewer 11: @jeffvance I am not suggesting it as the ultimate configuration, as it won't be portable, but I want COSI to give the driver's code (through grpc) all the BR/BAR information (including custom labels) - so it will be the provider's decision if to lose some portability in favor of supporting something that the COSI API has not yet incorporated into BR/BAR spec, where creating a Class for every request doesn't make sense. 

For example a COSI-driver that only imports existing cloud buckets - the BR will have to specify the external bucket name to import for it to be usable for users without admins creating classes - yes, not portable to another provider, but it can be portable between clouds and be a solid solution for many simple use cases.

Should COSI really be opinionated to prevent this?? Portability is important but also being open to customization,  at least until the COSI spec will cover these custom cases in a portable way :)

Pulling @saad-ali @wlan0 @bswartz for your opinions

Author: I'm open to input from others. My understanding of what you're suggesting is for COSI to support passing user-entered values to the driver. My opinion so far is that this could be used to circumvent bucket settings defined in the BucketClass by the admin. I also see value in allowing the user to define certain things and not have to bother the admin to create new BCs which are almost identical just to support some object store features. 

In a much earlier version of this KEP, we proposed a way to do both: give admin the ultimate say in what is passed to the driver, but allow the user some control. We did this by adding 2 fields to the BC:
.  enforcedConfig: [string]string map
.  defaultConfig:    [string]string map

and one field to the BucketRequest:
.   bucketConfig:  [string]string map

where _enforcedConfig_ could not be overwritten or bypassed, and _bucketConfig_ was merged into (overwriting) _defaultConfig_. This basic idea was rejected, in part, due to portability concerns.

Using labels or annotations as an "informal" way to provide the same feature doesn't remove the concern AFAICT.

Reviewer 11: I don't think we should standardize a way to bypass our proper COSI design.
But I think we should allow COSI drivers to decide if they need more information from the user, and not prevent it.

Author: Today in the KEP the driver received _bucket_context_, which is `Bucket.parameters`. COSI can control what's in this field but it's currently the BC's parameters. It sounds like you'd like the driver to be passed in an additional field, something like _user_config_ or _user_parameters_ so the driver sees this field as being different from what's defined in the BC. 

Interested in input from others.

1. `finalizers`: added by COSI to defer `BucketRequest` deletion until backend deletion succeeds.
1. `protocol`: (required) specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.

> 1. `bucketPrefix`: (optional) prefix prepended to a COSI-generated new bucket name, eg. “yosemite-photos-". If `bucketInstanceName` is supplied then `bucketPrefix` is ignored because the request is for access to an existing bucket.

Reviewer 11: Is this prefix applied to the Bucket CR name? or the backend bucket name for greenfield case?

Author: No. The BR.name is up to the user -- it is their BR. The prefix is applied to greenfield backend bucket names.

Reviewer 11: So if a user creates a greenfield BR without specifying a BR.spec.bucketInstanceName - what will be the B.name?

Author: BR.bucketInstanceName is the name of the cluster-scoped `Bucket` instance. If the user fills in this field then cosi knows we have brownfield (a new bucket will not be created). If B was created by the admin then its name is set by the admin. If cosi generated the B, then its name is non-deterministic; however, in this case the user who created the greenfield BR also creates a BAR that references their BR -- B's name is unimportant in this case.

Reviewer 11: Got it, my only comment is that I would still prefer to define what is the generated B.name going to be rather than leave it unspecified.

Author: B.name, when generated by cosi, is defined in the kep as:
```
When created by COSI, the Bucket name is generated in this format: "<BR.namespace*>-<BR.name*>-". If an admin creates a Bucket, as is necessary for brownfield access, they can use any name. The uuid is unique within a cluster. * the first 10 characters of the namespace and the name are used.
```
Is this what you're looking for?

Reviewer 11: @jeffvance thanks yes, I missed it.
I added a comment there (https://github.com/kubernetes/enhancements/pull/1383/files#r483814657).

For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.

Since there is a 1:1 mapping between a BR and a `Bucket`, when multiple BRs request access to the same backend bucket, each separate `Bucket` points to the same backend bucket.

```
apiVersion: cosi.io/v1alpha1
kind: Bucket
metadata:
  name: [1]
  labels:
    cosi.io/provisioner: [2]
  finalizers:
    - cosi.io/finalizer [3]
spec:
  provisioner: [4]
  retentionPolicy: [5]
  anonymousAccessMode: [6]
  bucketClassName: [7]
  bucketRequest: [8]
    name:
    namespace:
  bucketAccessName: [9]
  protocol: [10]
    protocolSignature: ""
    azureBlob:
      containerName:
      storageAccount:
    s3:
      version:
      endpoint:
      bucketName:
      region:
      signatureVersion:
    gcs:
      bucketName:
      privateKeyName:
      projectId:
      serviceAccount:

>   allowedNamespaces: [11]

Reviewer 11: Is allowedNamespaces still relevant now that every namespace has its own cluster-wide Bucket instance?

Author: Yes. _allowedNamespaces_ is defined in the BC and copied to the `Bucket` instance when COSI provisions a new backend bucket. This field lets an admin control which namespaces are allowed to request the creation of a bucket, or to request access to a bucket. A while back it was dropped from the KEP but community feedback was to add it back. As a note: a BC is immutable (like true for storage classes) but a `Bucket` instance can be mutated by the admin. So, _allowedNamespace_ in the `Bucket` could be changed to support additional namespaces or to remove a ns.

Reviewer 11: I understand what it was meant to be before. But now that we changed B to be 1-1 with BR per user/app namespace, then what is the meaning of having allowedNamespace in every B that is off sync between all the B's that refer to the same backend bucket?

Author: @guymguym I understand your question better now: once we have a B it implicitly means that the user was allowed to create their BR in their ns and no other namespaces will using that B. Good point! I think we only need _allowedNamespaces_ in the BC and not replicated to the B.

Reviewer 11: Continued in https://github.com/kubernetes/enhancements/pull/1383/files#r483796754

For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.

Since there is a 1:1 mapping between a BR and a `Bucket`, when multiple BRs request access to the same backend bucket, each separate `Bucket` points to the same backend bucket.

```yaml
apiVersion: cosi.io/v1alpha1
kind: Bucket
metadata:
  name: [1]
  labels: [2]
    cosi.io/provisioner:
    cosi.io/bucket-prefix:
  finalizers:
    - cosi.io/finalizer [3]
spec:
  provisioner: [4]
  retentionPolicy: [5]
  anonymousAccessMode: [6]
  bucketClassName: [7]
  bucketRequest: [8]
    name:
    namespace:
  allowedNamespaces: [9]
    - name:
  protocol: [10]
    protocolSignature: ""
    azureBlob:
      containerName:
      storageAccount:
    s3:
      version:
      endpoint:
      bucketName:
      region:
      signatureVersion:
    gcs:
      bucketName:
      privateKeyName:
      projectId:
      serviceAccount:
    - name:
  parameters: [11]
status:
  bucketAvailable: [12]
```

> 1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"<BR.namespace*>-<BR.name*>-<uuid>"_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name. The uuid is unique within a cluster. `*` the first 10 characters of the namespace and the name are used.

Reviewer 11: @jeffvance 10 first characters?? this is way too little. Think about "openshift-" which is already 10 characters long....
Name limit is 253 and uuid is 36. we can limit to 100 first characters and still be on the safe side, no?

Author: uuid = 36. `metadata.name` and ns fields are 256 each. So we have 256-36-2 (2 hyphens) avail. So looks like we can use up to 109 chars from ns and from name.

+ If the retention policy is "Retain" then the `Bucket` remains "Released" and it can potentially be reused.
+ When COSI sees the `Bucket` is "Deleted", it removes all finalizers and the resources are garbage collected.

## Grant Bucket Access
This workflow describes the automation supporting granting access to an existing backend bucket.

Here is the workflow:

+ The admin creates a cluster-scoped `Bucket` representing the backend bucket.
+ The user creates a `BucketRequest` referencing this `Bucket` instance and COSI follows the "Create Bucket" workflow.
+ The user creates a `BucketAccessRequest` (BAR) which references their BR.
+ COSI central controller sees a new `BucketAccessRequest` (BAR) and gets the `BAR.BucketAccessClass` (BAC).
+ COSI creates the cluster-scoped `BucketAccess` (BA) filling in the field from the BAR and BAC.
+ The sidecar sees the new BA and gRPC calls the driver to grant access to the backend bucket.
+ COSI updates the BAR and BA to reference each other.
+ The cosi-node-adapter, which has been waiting for the BAR to be processed, writes access tokens to the app pod's specified mount point.

> ## Revoke Bucket Access

Reviewer 11: In the last meeting we discussed Admin forcing access revocation by deleting the BA.
@jeffvance Should we add this flow?

Author: Yes. I think it's worth describing what we expect when the admin deletes the B or BA (assuming their finalizers are still in place).

+ _static provisioning_ - custom resource creation is done by the admin rather than via COSI automation. This may also include _driverless_ but they are independent.

# Proposal

## APIs

### Storage APIs

#### BucketRequest

A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.

There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketRequest
metadata:
  name:
  namespace:
  labels:
    cosi.io/provisioner: [1]
  finalizers:
    - cosi.io/finalizer [2]
spec:
  protocol: [3]
  bucketPrefix: [4]
  bucketClassName: [5]
  bucketInstanceName: [6]
status:
  bucketAvailable: [7]
```

> 1. `labels`: added by COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.

Reviewer 11: Another note on this same line - I think it's worth noting here the rational of having the provisioner name as a label so that it is easy to filter out requests by the provisioner for admin commands - `kubectl get buckets -l cosi.io/provisioner=...`

Author: or `kubectl get all -l cosi.io/provisioner=...`

11. `parameters`: a copy of the BucketClass parameters.
12. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.

#### BucketClass

A cluster-scoped custom resource to provide admins control over the handling of bucket provisioning.  The `BucketClass` (BC) defines a retention policy, driver specific parameters, and the provisioner name. A list of allowed namespaces can be specified to restrict new bucket creation and access to existing buckets. A default bucket class can be defined for each supported protocol. This allows the bucket class to be omitted from a `BucketRequest`. All of the `BucketClass` fields are copied to the `Bucket` instance.

> Note: like storage classes, a `BucketClass` is immutable, but a `Bucket` is mutable and contains all of the BC fields.

```yaml
apiVersion: cosi.io/v1alpha1
kind: BucketClass
metadata:
  name: 
provisioner: [1]
isDefaultBucketClass: [2]
protocol: {"azureblob", "gcs", "s3", ... } [3]
anonymousAccessMode: [4]
retentionPolicy: {"Delete", "Retain"} [5]
allowedNamespaces: [6]
  - name:
parameters: [7]
```

> 1. `provisioner`: the name of the driver. If supplied the driver container and sidecar container are expected to be deployed. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".

Reviewer 11: Is this "ns/name" format required - Is COSI going to parse it in order to locate the provisioner namespace?
If so it might be better to split into 2 fields?

Author: We are proposing the same as implemented by CSI which brings familiarity. I do agree that two fields are clearer.

2. `isDefaultBucketClass`: (optional) boolean, default is false. If set to true then a `BucketRequest` may not need to specify a `BucketClass`. If the greenfield `BucketRequest` omits the `BucketClass` and a default `BucketClass`'s protocol matches the `BucketRequest`'s protocol then the default bucket class is used; otherwise an error is logged.

> 3. `protocol`: (required) protocol supported by the associated object store. This field validates that the `BucketRequest`'s desired protocol is supported.

Reviewer 11: Is this a fixed enum including just s3/blob/gcs? 
or is this a free form string that COSI just matches between the BR and the BC?
I mean - can a provider decide to support more variants like "s3:v2" or "s3:ceph"?

Author: Good question! It is proposed to be a simple string match, but version has come up in reviews and has not been reflected in the KEP other than in B.protocol.s3.vesion. I think we need to support something like you've described. 
@wlan0 @brahmaroutu

Reviewer 11: Right, we discussed at least the S3 apiVersion -
https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/locking-api-versions.html

      bucketName:
      region:
      signatureVersion:

>       publicAccess:

Author: Is public not an attribute across all protocols?

1. `bucketClassName`: Name of the associated `BucketClass`.
1. `supportedProtocols`:  String array of protocols (e.g. s3, gcs, swift, etc.) supported by the associated object store.
1. `releasePolicy`: the release policy defined in the associated BucketClass (see [BucketClass](#BucketClass) for more information).

> 1. `bucketRef`: the name & namespace of the associated `Bucket`. For both brownfield cases, `bucketRef` is set by an admin and names the bucket to be accessed. For greenfield, it is added by the COSI controller.

Reviewer 3: In Brownfield, the COSI controller can still manage binding operations.  I know that pre-binding was a thing in k8s before dynamic provisioning, but I think in hindsight, it wasn't regarded as a great solution.  Binding and secret copying, under brownfield, can be completely automated.

Author: Agree since I assume by _binding_ I think we mean that `Bucket.status.bucketContentName` == `BucketContent.name` && `BucketContent.bucketRef.name` ==  `Bucket.name`

Reviewer 3: Right, that's part of it, but those are effectively meaningless if the Secret isn't provided to the user via automation.

    - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed
1. `accessMode`: Declares the level of access given to credentials provisioned through this class.     If empty, drivers may set defaults.
1. `bucketClassName`: Name of the associated `BucketClass`.

> 1. `bindableNamespaces`: An array of namespaces, identified by a name and uid, that are permitted to bind Buckets to this BucketContent.  Provided to allow admins a layer of control in who can access the backing bucket.  Does **not** reflect or alter the backing bucket’s ACLs or access policies.

Author: Is this required? If omitted is the default `*`?

isDefaultBucketClass: [2]
supportedProtocols: {"azureblob", "gcs", "s3", ... } [3]
accessMode: {"ro", "wo", "rw"} [4]

> private: boolean [5]

Author: _isPublic_ is missing from the BucketClass

1. `isDefaultBucketClass`: boolean. When true, signals that the controller should attempt to match `Bucket`’s without a defined `BucketClass` to this class, accounting for the `Bucket`’s requested protocol.  Multiple default classes for the same protocol will produce undefined behaviour.
1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses.
1. `accessMode`: (Optional) ACL setting specifying the default accessibility of .

> 1. `isPublic`: ACL setting specifying if the a generated storage instance should be publicly accessible.  Common setting among protocols.

Author: this field is missing in the spec
